{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Loading data and working with grandPy objects\n",
    "\n",
    "<span style=\"color:green\">Notiz: alles was rot ist, ist noch nicht umgesetzt oder müsste noch überarbeitet werden durch Links etc.</span>\n",
    "\n",
    "GrandPy is a Python package for the analysis of RNA-seq experiments involving metabolic RNA labeling with nucleotide conversion, such as SLAM-seq experiments [[1]](https://www.nature.com/articles/nmeth.4435). In such experiments, nucleoside analogs such as 4sU are added to living cells, which take it up and incorporate it into newly synthesized RNA. Before sequencing, 4sU is converted into a cytosin analog. Reads covering 4sU sites therefore have characteristic T-to-C mismatches after read mapping, in principle providing the opportunity to differentiate newly synthesized (during the time of labeling) from preexisting RNA.\n",
    "\n",
    "Confounders such as sequencing errors or read that originate from newly synthesized RNA but, by chance, do not cover sites of 4sU incorporation (usually 20-80% of all \"new read\") can be handled using specialized methods such as GRAND-SLAM [[2]](https://academic.oup.com/bioinformatics/article/34/13/i218/5045735?login=true).\n",
    "\n",
    "# Reading in the data\n",
    "\n",
    "Throughout this vignette, we will be using the GRAND-SLAM processed SLAM-seq data set from Finkel et al.2021 [[3]](https://academic.oup.com/bioinformatics/article/34/13/i218/5045735). The data set contains time series (progressive labeling) samples from a human epithelial cell line (Calu3 cells); half of the samples were infected with SARS-CoV-2 for different periods of time.\n",
    "\n",
    "The output of GRAND-SLAM is a tsv file where rows are genes and columns are read counts and other statistics (e.g., the new-to-total RNA ratio) for all samples. The data set is available on zenodo (\"https://zenodo.org/record/5834034/files/sars.tsv.gz\"). We start by reading this file into Python:"
   ],
   "id": "52df091142cc2d28"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:55.348433Z",
     "start_time": "2025-06-22T13:50:54.846684Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Package installation\n",
    "from Py.load import *\n",
    "sars = read_grand(\"../data/sars.tsv\", design=(\"Condition\", \"Time\", \"Replicate\"))\n",
    "print(sars.columns)"
   ],
   "id": "ce3f00b85976da6",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected dense format -> using dense reader\n",
      "['Mock.no4sU.A', 'Mock.1h.A', 'Mock.2h.A', 'Mock.2h.B', 'Mock.3h.A', 'Mock.4h.A', 'SARS.no4sU.A', 'SARS.1h.A', 'SARS.2h.A', 'SARS.2h.B', 'SARS.3h.A', 'SARS.4h.A']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\PycharmProjects\\grandpy\\Py\\utils.py:61: UserWarning: 17 Duplicate gene symbols found: STPG4, PDE11A, TXNRD3NB, ARL14EPL, HIST1H3D (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\n",
      "  warnings.warn(f\"{len(duplicates_list)} Duplicate gene symbols found: {', '.join(duplicates_list[:5])} (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\")\n"
     ]
    }
   ],
   "execution_count": 1
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "When reading in the file, we have to define the ``design`` vector. This is used to infer metadata automatically from sample names. Here sample names consist of three parts separated by dots as shown above (the Columns function returns the sample names or cell ids when analyzing a single cell data set). Each part in the sample name represents an aspect of the design. For example, the sample named Mock.2h.A is a sample from the mock condition (i.e. not infected by SARS-CoV-2), subjected to metabolic labeling for 2 hours, and is the first replicate (i.e. replicate \"A\"). This sample name is consistent with the three element design vector used above. It is possible to specify other design elements (of course the samples would have to be named accordingly). A list of reasonable options is predefined in the dictionary `DESIGN_KEYS`.\n",
    "\n",
    "There are names (i.e. the things you specify in the design vector) that have additional semantics. For example, for the name `duration.4sU`the values are interpreted like this: 4h is converted into the number 4, 30min into 0.5, and no4sU into 0. For more information, see [below](#column-metadata)). The design vector is mandatory. Attempting to read in the data without it results in an error:"
   ],
   "id": "2b9995e1b21a3782"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:56.784515Z",
     "start_time": "2025-06-22T13:50:56.308093Z"
    }
   },
   "cell_type": "code",
   "source": "sars_wrong = read_grand(\"../data/sars.tsv\") # Fehlermeldung könnte man noch verkürzen",
   "id": "bc7af08f0cc34232",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected dense format -> using dense reader\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\PycharmProjects\\grandpy\\Py\\utils.py:61: UserWarning: 17 Duplicate gene symbols found: STPG4, PDE11A, TXNRD3NB, ARL14EPL, HIST1H3D (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\n",
      "  warnings.warn(f\"{len(duplicates_list)} Duplicate gene symbols found: {', '.join(duplicates_list[:5])} (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\")\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Design must be specified.",
     "output_type": "error",
     "traceback": [
      "\u001B[31m---------------------------------------------------------------------------\u001B[39m",
      "\u001B[31mValueError\u001B[39m                                Traceback (most recent call last)",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[2]\u001B[39m\u001B[32m, line 1\u001B[39m\n\u001B[32m----> \u001B[39m\u001B[32m1\u001B[39m sars_wrong = \u001B[43mread_grand\u001B[49m\u001B[43m(\u001B[49m\u001B[33;43m\"\u001B[39;49m\u001B[33;43m../data/sars.tsv\u001B[39;49m\u001B[33;43m\"\u001B[39;49m\u001B[43m)\u001B[49m \u001B[38;5;66;03m# Fehlermeldung könnte man noch verkürzen\u001B[39;00m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\PycharmProjects\\grandpy\\Py\\load.py:649\u001B[39m, in \u001B[36mread_grand\u001B[39m\u001B[34m(prefix, pseudobulk, targets, **kwargs)\u001B[39m\n\u001B[32m    647\u001B[39m file_path = resolve_prefix_path(prefix, pseudobulk=pseudobulk, targets=targets)\n\u001B[32m    648\u001B[39m \u001B[38;5;28mprint\u001B[39m(\u001B[33m\"\u001B[39m\u001B[33mDetected dense format -> using dense reader\u001B[39m\u001B[33m\"\u001B[39m)\n\u001B[32m--> \u001B[39m\u001B[32m649\u001B[39m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mread_dense\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;28;43mstr\u001B[39;49m\u001B[43m(\u001B[49m\u001B[43mfile_path\u001B[49m\u001B[43m)\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43m*\u001B[49m\u001B[43m*\u001B[49m\u001B[43mkwargs\u001B[49m\u001B[43m)\u001B[49m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\PycharmProjects\\grandpy\\Py\\load.py:573\u001B[39m, in \u001B[36mread_dense\u001B[39m\u001B[34m(file_path, default_slot, design, classification_genes, classification_genes_label, classify_genes_func)\u001B[39m\n\u001B[32m    543\u001B[39m \u001B[38;5;28;01mdef\u001B[39;00m\u001B[38;5;250m \u001B[39m\u001B[34mread_dense\u001B[39m(file_path, default_slot=\u001B[33m\"\u001B[39m\u001B[33mcount\u001B[39m\u001B[33m\"\u001B[39m, design=\u001B[38;5;28;01mNone\u001B[39;00m, *, classification_genes=\u001B[38;5;28;01mNone\u001B[39;00m, classification_genes_label=\u001B[33m\"\u001B[39m\u001B[33mViral\u001B[39m\u001B[33m\"\u001B[39m, classify_genes_func=\u001B[38;5;28;01mNone\u001B[39;00m):\n\u001B[32m    544\u001B[39m \u001B[38;5;250m    \u001B[39m\u001B[33;03m\"\"\"\u001B[39;00m\n\u001B[32m    545\u001B[39m \u001B[33;03m    Reads a GRAND-SLAM TSV file as dense (NumPy) matrices and returns a GrandPy object.\u001B[39;00m\n\u001B[32m    546\u001B[39m \n\u001B[32m   (...)\u001B[39m\u001B[32m    570\u001B[39m \u001B[33;03m        A GrandPy object populated with dense matrices and metadata.\u001B[39;00m\n\u001B[32m    571\u001B[39m \u001B[33;03m    \"\"\"\u001B[39;00m\n\u001B[32m--> \u001B[39m\u001B[32m573\u001B[39m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43m_read\u001B[49m\u001B[43m(\u001B[49m\u001B[43mfile_path\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43msparse\u001B[49m\u001B[43m=\u001B[49m\u001B[38;5;28;43;01mFalse\u001B[39;49;00m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdefault_slot\u001B[49m\u001B[43m=\u001B[49m\u001B[43mdefault_slot\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdesign\u001B[49m\u001B[43m=\u001B[49m\u001B[43mdesign\u001B[49m\u001B[43m,\u001B[49m\n\u001B[32m    574\u001B[39m \u001B[43m                 \u001B[49m\u001B[43mclassification_genes\u001B[49m\u001B[43m=\u001B[49m\u001B[43mclassification_genes\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mclassification_genes_label\u001B[49m\u001B[43m=\u001B[49m\u001B[43mclassification_genes_label\u001B[49m\u001B[43m,\u001B[49m\n\u001B[32m    575\u001B[39m \u001B[43m                 \u001B[49m\u001B[43mclassify_genes_func\u001B[49m\u001B[43m=\u001B[49m\u001B[43mclassify_genes_func\u001B[49m\u001B[43m)\u001B[49m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\PycharmProjects\\grandpy\\Py\\load.py:761\u001B[39m, in \u001B[36m_read\u001B[39m\u001B[34m(file_path, sparse, default_slot, design, classification_genes, classification_genes_label, classify_genes_func, pseudobulk, targets)\u001B[39m\n\u001B[32m    758\u001B[39m     classify_genes_func = \u001B[38;5;28;01mlambda\u001B[39;00m gene_info: classify_genes(gene_info, custom_classes=custom, use_default=\u001B[38;5;28;01mTrue\u001B[39;00m)\n\u001B[32m    760\u001B[39m gene_info = build_gene_info(df, classify_genes_func)\n\u001B[32m--> \u001B[39m\u001B[32m761\u001B[39m coldata = \u001B[43mbuild_coldata\u001B[49m\u001B[43m(\u001B[49m\u001B[43msample_names\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdesign\u001B[49m\u001B[43m)\u001B[49m\n\u001B[32m    762\u001B[39m slots = pad_slots(slots, sparse, coldata, slot_sample_names)\n\u001B[32m    764\u001B[39m metadata = {\n\u001B[32m    765\u001B[39m     \u001B[33m\"\u001B[39m\u001B[33mDescription\u001B[39m\u001B[33m\"\u001B[39m: \u001B[33m\"\u001B[39m\u001B[33mLoaded via read_grand() (TSV)\u001B[39m\u001B[33m\"\u001B[39m,\n\u001B[32m    766\u001B[39m     \u001B[33m\"\u001B[39m\u001B[33mdefault_slot\u001B[39m\u001B[33m\"\u001B[39m: default_slot,\n\u001B[32m   (...)\u001B[39m\u001B[32m    770\u001B[39m     \u001B[33m\"\u001B[39m\u001B[33mtargets\u001B[39m\u001B[33m\"\u001B[39m: targets\n\u001B[32m    771\u001B[39m }\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\PycharmProjects\\grandpy\\Py\\load.py:267\u001B[39m, in \u001B[36mbuild_coldata\u001B[39m\u001B[34m(names, design)\u001B[39m\n\u001B[32m    264\u001B[39m max_fields = \u001B[38;5;28mmax\u001B[39m(\u001B[38;5;28mlen\u001B[39m(parts) \u001B[38;5;28;01mfor\u001B[39;00m parts \u001B[38;5;129;01min\u001B[39;00m split_names)\n\u001B[32m    266\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m design \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[32m--> \u001B[39m\u001B[32m267\u001B[39m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mValueError\u001B[39;00m(\u001B[33m\"\u001B[39m\u001B[33mDesign must be specified.\u001B[39m\u001B[33m\"\u001B[39m)\n\u001B[32m    268\u001B[39m     \u001B[38;5;66;03m# predefined = list(DESIGN_KEYS.values())\u001B[39;00m\n\u001B[32m    269\u001B[39m     \u001B[38;5;66;03m# design = tuple(predefined[i] if i < len(predefined) else f\"Design_{i+1}\" for i in range(max_fields))\u001B[39;00m\n\u001B[32m    270\u001B[39m \u001B[38;5;28;01melif\u001B[39;00m \u001B[38;5;28mlen\u001B[39m(design) < max_fields:\n",
      "\u001B[31mValueError\u001B[39m: Design must be specified."
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Alternatively, a table containing the metadata can be specified. Make sure that it contains a `Name` column matching the names in the GRAND-SLAM output table:",
   "id": "6d8012805e46d830"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "name = [\"Mock.no4sU.A\", \"Mock.1h.A\", \"Mock.2h.A\", \"Mock.2h.B\", \"Mock.3h.A\", \"Mock.4h.A\", \"SARS.no4sU.A\", \"SARS.1h.A\", \"SARS.2h.A\", \"SARS.2h.B\", \"SARS.3h.A\", \"SARS.4h.A\"]\n",
    "\n",
    "conditions = [\"Mock\"] * 6 + [\"SARS\"] * 6\n",
    "\n",
    "design_df = pd.DataFrame({\"Name\": name,\n",
    "                          \"Condition\": conditions})\n",
    "\n",
    "sars_meta = read_grand(\"../data/sars.tsv\", design = design_df)"
   ],
   "id": "7e52409fff9b9aee",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# What is the grandPy object\n",
    "\n",
    "`read_grand` returns a grandPy object, which contains\n",
    "\n",
    "1. metadata for genes\n",
    "2. metadata for samples/cells (as inferred from the sample names by the design parameter)\n",
    "3. all data matrices (counts, normalized counts, ntrs, etc. these types of data are called \"slots\")\n",
    "4. analysis results\n",
    "\n",
    "Metadata (1. and 2.) are described below. How to work with the data matrices and analysis results is described in a separate [vignette](working_with_data_matrices_and_analysis_results.ipynb).\n",
    "\n",
    "# Working with grandPy objects\n",
    "\n",
    "Here we will see how to work with grandPy objects in general. A short summary can be displayed when `print`ing the object, and there are several functinos to retrieve general information about the object:"
   ],
   "id": "898d70860c82c871"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:57.047459Z",
     "start_time": "2025-06-22T13:50:57.045076Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars) # Link fehlt",
   "id": "868a9077ac36ab95",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GrandPy:\n",
      "Read from sars\n",
      "19659 genes, 12 samples/cells\n",
      "Available data slots: ['count', 'ntr', 'alpha', 'beta']\n",
      "Available analyses: []\n",
      "Available plots: {}\n",
      "Default data slot: count\n",
      "\n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:57.211920Z",
     "start_time": "2025-06-22T13:50:57.209490Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.title)",
   "id": "cbdc3f0edab03e52",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sars\n"
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:57.373677Z",
     "start_time": "2025-06-22T13:50:57.369996Z"
    }
   },
   "cell_type": "code",
   "source": "print(len(sars.genes)) # so?",
   "id": "3d3886922a51a5be",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19659\n"
     ]
    }
   ],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:57.533735Z",
     "start_time": "2025-06-22T13:50:57.531117Z"
    }
   },
   "cell_type": "code",
   "source": "print(len(sars.coldata)) # so?",
   "id": "443d3ee445a9d8a2",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n"
     ]
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "It is straight-forward to filter genes:",
   "id": "20c9e3e4ad8443d8"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:57.851294Z",
     "start_time": "2025-06-22T13:50:57.849228Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# sars = sars.filter_genes # fehlt bei processing.py\n",
    "# print(len(sars.genes))"
   ],
   "id": "247da2955b9c3e14",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "By default genes are retained if they have 100 read counts in at least half of the samples (or cells). There are many options how to filter by genes (note that `filter_genes` returns a new grandPy object, and below we directly call `len()` on this new object to check how many genes are retained by filtering):",
   "id": "2956390d4b69d53a"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:58.217034Z",
     "start_time": "2025-06-22T13:50:58.214623Z"
    }
   },
   "cell_type": "code",
   "source": "# print(\"Genes with at least 1000 read counts in half of the columns: \\n\", len(sars.genes)) # filter_genes fehlt bei processing.py",
   "id": "2ca2127594f21a4e",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:59.066203Z",
     "start_time": "2025-06-22T13:50:59.064326Z"
    }
   },
   "cell_type": "code",
   "source": "# print(\"Genes with at least 1000 read counts in half of the columns (retain two genes that are otherwise filtered): \\n\", ) # filter_genes fehlt bei processing.py",
   "id": "5b961114b3c883d4",
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:59.272777Z",
     "start_time": "2025-06-22T13:50:59.270339Z"
    }
   },
   "cell_type": "code",
   "source": "# print(\"Keep only these two genes: \\n\", ...) # filter_genes fehlt bei processing.py",
   "id": "c3fd30fb01e7e6c",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:59.407343Z",
     "start_time": "2025-06-22T13:50:59.405318Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# sars = sars.normalizeTPM # normalizeTPM() fehlt bei processing.py\n",
    "# print(\"Genes with at least 10 TPM in half of the columns: \\n\", ...)"
   ],
   "id": "9f63268ac360d0c4",
   "outputs": [],
   "execution_count": 11
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "`filter_genes()` essentially removes rows from the data slots. It is also possible to remove columns (i.e. samples or cells). This is done using the subset function:",
   "id": "749285f3714131ec"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:50:59.779559Z",
     "start_time": "2025-06-22T13:50:59.770039Z"
    }
   },
   "cell_type": "code",
   "source": [
    "mock = sars[:, sars.coldata[\"Condition\"] == \"Mock\"]\n",
    "print(mock) # filter_genes() fehlt noch, daher genes Anzhal noch falsch aber samples/cells richtig"
   ],
   "id": "1a3bd6c206bf4b9c",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GrandPy:\n",
      "Read from sars\n",
      "19659 genes, 6 samples/cells\n",
      "Available data slots: ['count', 'ntr', 'alpha', 'beta']\n",
      "Available analyses: []\n",
      "Available plots: {}\n",
      "Default data slot: count\n",
      "\n"
     ]
    }
   ],
   "execution_count": 12
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "The new grandPy object now only has 6 columns. The `columns`parameter to subset must be a logical vector, and you can use the names of the column metadata table (see below) as variables (i.e. the parameter here is a logical vector with all samples being TRUE where the `Condition`column is equal to \"Mock\".\n",
    "\n",
    "A closely related function is `split`, which returns a list of several grandPy objects, each composed of samples having the same `Condition`."
   ],
   "id": "fa0b3b984127293b"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:00.132235Z",
     "start_time": "2025-06-22T13:51:00.130266Z"
    }
   },
   "cell_type": "code",
   "source": "# .split() fehlt in grandpy.py",
   "id": "46ea376976435f49",
   "outputs": [],
   "execution_count": 13
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:00.319248Z",
     "start_time": "2025-06-22T13:51:00.317172Z"
    }
   },
   "cell_type": "code",
   "source": "# lapply - funktion raussuchen in python",
   "id": "80f6bf6d81e03ef9",
   "outputs": [],
   "execution_count": 14
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "The inverse of `split` is `merge`:",
   "id": "4ba5e07762570b05"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:00.702315Z",
     "start_time": "2025-06-22T13:51:00.699556Z"
    }
   },
   "cell_type": "code",
   "source": "#",
   "id": "3544a037a12253d9",
   "outputs": [],
   "execution_count": 15
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Note that we merged such that now we have first the SARS samples and then the Mock samples. We can reorder by slightly abusing `subset` (note that we actually do not omit any columns, but just define a different order):",
   "id": "cdeabc781b9d0b7f"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:01.089872Z",
     "start_time": "2025-06-22T13:51:01.087843Z"
    }
   },
   "cell_type": "code",
   "source": "#",
   "id": "9eb8ab4e33534bea",
   "outputs": [],
   "execution_count": 16
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Gene metadata\n",
    "\n",
    "Here we see how to work with metadata for genes. The gene metadata essentially is a table that can be retrieved using the `gene_info` function:"
   ],
   "id": "7718ced5527c771"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:01.494748Z",
     "start_time": "2025-06-22T13:51:01.489729Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.gene_info.head(10))",
   "id": "461900bf1f05bf92",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                  Symbol             Gene  Length      Type\n",
      "Symbol                                                     \n",
      "HNRNPCL3        HNRNPCL3  ENSG00000277058    1148  Cellular\n",
      "AL137802.1    AL137802.1  ENSG00000224174     327  Cellular\n",
      "AL137798.2    AL137798.2  ENSG00000282143     817  Cellular\n",
      "PLA2G2F          PLA2G2F  ENSG00000158786    4129  Cellular\n",
      "Z98257.1        Z98257.1  ENSG00000227066    2493  Cellular\n",
      "UBXN10            UBXN10  ENSG00000162543    5571  Cellular\n",
      "Z98257.2        Z98257.2  ENSG00000284710    1548  Cellular\n",
      "ACTL8              ACTL8  ENSG00000117148    1861  Cellular\n",
      "MIR1302-2HG  MIR1302-2HG  ENSG00000243485     712  Cellular\n",
      "NBPF3              NBPF3  ENSG00000142794    4401  Cellular\n"
     ]
    }
   ],
   "execution_count": 17
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Each gene has associated gene ids and symbols. Gene ids and symbols as well as the transcript length are part of GRAND-SLAM output. The `type` column is inferred automatically (see below).\n",
    "\n",
    "Genes can be identified by the `genes` function:"
   ],
   "id": "6fa264923989ae12"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:01.895166Z",
     "start_time": "2025-06-22T13:51:01.891835Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.genes[:20])",
   "id": "604f34be22540199",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HNRNPCL3', 'AL137802.1', 'AL137798.2', 'PLA2G2F', 'Z98257.1', 'UBXN10', 'Z98257.2', 'ACTL8', 'MIR1302-2HG', 'NBPF3', 'ALPL', 'LDLRAD2', 'DISP3', 'FBXO44', 'FBXO6', 'DRAXIN', 'PTPRU', 'CDA', 'PINK1', 'SMIM1']\n"
     ]
    }
   ],
   "execution_count": 18
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:02.108361Z",
     "start_time": "2025-06-22T13:51:02.102754Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# head(Genes(sars,use.symbols = FALSE), n=20)      # the first 20 genes, but now use the ids\n",
    "# print(sars.genes[:20])"
   ],
   "id": "6bf9efffac404c05",
   "outputs": [],
   "execution_count": 19
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:02.306164Z",
     "start_time": "2025-06-22T13:51:02.303761Z"
    }
   },
   "cell_type": "code",
   "source": "# Genes(sars,genes = c(\"MYC\",\"ORF1ab\"),use.symbols = FALSE)    # convert to ids",
   "id": "2f579e835d6a98b2",
   "outputs": [],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:02.472036Z",
     "start_time": "2025-06-22T13:51:02.469710Z"
    }
   },
   "cell_type": "code",
   "source": "# Genes(sars,genes = \"YC\", regex = TRUE)           # retrieve all genes matching to the regular expression YC",
   "id": "ed41f6a04b3f70a3",
   "outputs": [],
   "execution_count": 21
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "During reading the data into R using `read_grand`, the `Type` column i9s inferred using the `classify_genes` function. By default, this will recognize mitochondrial genes (MT prefix of the gene symbol), ERCC spike-ins, and Ensembl gene identifiers (which it will call \"cellular\"). Here we also have the viral genes, which are not properly recognized:",
   "id": "fa9ff85f682e2584"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:02.826829Z",
     "start_time": "2025-06-22T13:51:02.822798Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.gene_info[\"Type\"].value_counts()) # Table?",
   "id": "970e928f7241b41c",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Type\n",
      "Cellular    19648\n",
      "Unknown        11\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "execution_count": 22
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "If you want to define your own types, you can do this easily by specifying the `classify_genes` parameter when read in your data:",
   "id": "aa2cb42b2d41ab19"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:03.358243Z",
     "start_time": "2025-06-22T13:51:03.224978Z"
    }
   },
   "cell_type": "code",
   "source": [
    "viral_genes = ['ORF3a','E','M','ORF6','ORF7a','ORF7b','ORF8','N','ORF10','ORF1ab','S']\n",
    "sars = read_grand(\"../data/sars.tsv\", design=(\"Condition\", \"Time\", \"Replicate\"), classification_genes=viral_genes)\n",
    "\n",
    "print(sars.gene_info[\"Type\"].value_counts())"
   ],
   "id": "f4b3c6723f5d65bf",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected dense format -> using dense reader\n",
      "Type\n",
      "Cellular    19648\n",
      "Viral          11\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\PycharmProjects\\grandpy\\Py\\utils.py:61: UserWarning: 17 Duplicate gene symbols found: STPG4, PDE11A, TXNRD3NB, ARL14EPL, HIST1H3D (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\n",
      "  warnings.warn(f\"{len(duplicates_list)} Duplicate gene symbols found: {', '.join(duplicates_list[:5])} (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\")\n"
     ]
    }
   ],
   "execution_count": 23
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Note that each parameter to `classify_genes` must be named (`viral`) and must be a function that takes the gene metadata table and returns a logical vector.\n",
    "\n",
    "The `classify_genes` function has one additional important parameter, which defines how \"Unknown\" types are supposed to be called. For this data set, a similar behavior as aboce can be accomplished by:"
   ],
   "id": "dacfd31306c71151"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:03.777741Z",
     "start_time": "2025-06-22T13:51:03.645120Z"
    }
   },
   "cell_type": "code",
   "source": [
    "sars = read_grand(\"../data/sars.tsv\", design=(\"Condition\", \"Time\", \"Replicate\"), classification_genes=None, classification_genes_label=\"Unknown\",\n",
    "                 classify_genes_func=None)\n",
    "\n",
    "print(sars.gene_info[\"Type\"].value_counts())\n",
    "\n",
    "# das ist noch nicht ganz richtig hier"
   ],
   "id": "1df2d709f774619a",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected dense format -> using dense reader\n",
      "Type\n",
      "Cellular    19648\n",
      "Unknown        11\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\PycharmProjects\\grandpy\\Py\\utils.py:61: UserWarning: 17 Duplicate gene symbols found: STPG4, PDE11A, TXNRD3NB, ARL14EPL, HIST1H3D (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\n",
      "  warnings.warn(f\"{len(duplicates_list)} Duplicate gene symbols found: {', '.join(duplicates_list[:5])} (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\")\n"
     ]
    }
   ],
   "execution_count": 24
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "It is also straight-forward to add additional gene metadata:",
   "id": "2ff3b76ee181388c"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:04.147880Z",
     "start_time": "2025-06-22T13:51:04.145691Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# GeneInfo(sars,\"length.category\") <- cut(GeneInfo(sars,\"Length\"),\n",
    "#                                         breaks=c(0,2000,5000,Inf),\n",
    "#                                         labels = c(\"Short\",\"Medium\",\"Long\"))\n",
    "# table(GeneInfo(sars,\"length.category\"))"
   ],
   "id": "30fc82f69c008b8e",
   "outputs": [],
   "execution_count": 25
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Column metadata\n",
    "\n",
    "Samples for bulk experiments and cells in single cell experiments are in grandR jointly called \"columns\". The metadata for columns is a table that describes the experimental design we specified when reading in data in grandPy. It can be accessed via the `coldata` function. We can also see that the duration of 4sU has been interpreted and converted to a numeric value (compare \"Time\" with \"Time.original\")."
   ],
   "id": "56241a038792896c"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:05.156345Z",
     "start_time": "2025-06-22T13:51:05.151872Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.coldata)",
   "id": "a40906431d533df6",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      Name Condition  Time Replicate Time.original  no4sU\n",
      "Name                                                                     \n",
      "Mock.no4sU.A  Mock.no4sU.A      Mock   0.0         A         no4sU   True\n",
      "Mock.1h.A        Mock.1h.A      Mock   1.0         A            1h  False\n",
      "Mock.2h.A        Mock.2h.A      Mock   2.0         A            2h  False\n",
      "Mock.2h.B        Mock.2h.B      Mock   2.0         B            2h  False\n",
      "Mock.3h.A        Mock.3h.A      Mock   3.0         A            3h  False\n",
      "Mock.4h.A        Mock.4h.A      Mock   4.0         A            4h  False\n",
      "SARS.no4sU.A  SARS.no4sU.A      SARS   0.0         A         no4sU   True\n",
      "SARS.1h.A        SARS.1h.A      SARS   1.0         A            1h  False\n",
      "SARS.2h.A        SARS.2h.A      SARS   2.0         A            2h  False\n",
      "SARS.2h.B        SARS.2h.B      SARS   2.0         B            2h  False\n",
      "SARS.3h.A        SARS.3h.A      SARS   3.0         A            3h  False\n",
      "SARS.4h.A        SARS.4h.A      SARS   4.0         A            4h  False\n"
     ]
    }
   ],
   "execution_count": 26
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Additional semantics can also be defined, which is accomplished via the function `apply_design_semantics`, that generates a list for the `semantics` parameter of the function `build_coldata`, which in turn is used to infer metadata from sample names. We briefly explain these mechanisms with an example, but it is important to mention that in most cases, the desired metadata can be added after reading the data, as shown further below.\n",
    "\n",
    "<span style=\"color:red\">First, it is important to have a function that takes two parameters (a specific column of the original column of the original column metadata table + the name of this column) and returns a dataframe that is then `cbin`ed with the original column metadata table. There is one such predefined function in grandPy, which parses labeling durations:</span>"
   ],
   "id": "9ef9949fb6cfa4bb"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:06.389306Z",
     "start_time": "2025-06-22T13:51:06.384176Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from pandas import Series\n",
    "Series([\"5h\", \"30min\", \"no4sU\"]).map(parse_time_string)"
   ],
   "id": "f880daa19f81a7c9",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    5.0\n",
       "1    0.5\n",
       "2    0.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 27
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "We can easily define our own function like this:",
   "id": "e44a29c42585f3e8"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:07.288197Z",
     "start_time": "2025-06-22T13:51:07.282255Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def my_semantics_time(values, name):\n",
    "    df = pd.DataFrame({name: values})\n",
    "    df[name + \"_hr\"] = df[name].map(parse_time_string)\n",
    "    df[\"hpi\"] = df[name + \"_hr\"].apply(lambda x: f\"{x + 3}hpi\")\n",
    "    df.attrs[\"_semantics\"] = {name: \"time\"}\n",
    "    return df\n",
    "\n",
    "my_semantics_time([\"5h\", \"30min\", \"no4sU\"], \"Test\")"
   ],
   "id": "caaa8d9710463465",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    Test  Test_hr     hpi\n",
       "0     5h      5.0  8.0hpi\n",
       "1  30min      0.5  3.5hpi\n",
       "2  no4sU      0.0  3.0hpi"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Test</th>\n",
       "      <th>Test_hr</th>\n",
       "      <th>hpi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5h</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0hpi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30min</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.5hpi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>no4sU</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0hpi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 28
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "<span style=\"color:red\"> Here, it is important to mention that at 3h post infection, 4sU was added to the cells for 1, 2, 3 or 4h. The two no4sU samples are also 3h post infection. This function can now be used as `semantics`parameter for `read_grand` like this:</span>",
   "id": "697021a3ed1bbf0f"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:08.054207Z",
     "start_time": "2025-06-22T13:51:08.052164Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# sars.meta <- ReadGRAND(system.file(\"extdata\", \"sars.tsv.gz\", package = \"grandR\"),\n",
    "#                    design=function(names)\n",
    "#                      MakeColdata(names,\n",
    "#                                  c(\"Cell\",Design$dur.4sU,Design$Replicate),\n",
    "#                                  semantics=DesignSemantics(duration.4sU=my.semantics.time)\n",
    "#                                  ),\n",
    "#                  verbose=TRUE)"
   ],
   "id": "a3f28343d9b9eaf9",
   "outputs": [],
   "execution_count": 29
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "As mentioned above, it is in most cases easier to add additional metadata after loading.The infection time point can also be added by:",
   "id": "8f0614f9c546031c"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:08.703181Z",
     "start_time": "2025-06-22T13:51:08.698648Z"
    }
   },
   "cell_type": "code",
   "source": [
    "sars.coldata[\"hpi\"] = sars.coldata[\"Time\"].apply(lambda x: f\"{x + 3}hpi\")\n",
    "print(sars.coldata) # da es immutable ist, kann ich keine Spalten anfügen ..."
   ],
   "id": "525c52844f691e01",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      Name Condition  Time Replicate Time.original  no4sU\n",
      "Name                                                                     \n",
      "Mock.no4sU.A  Mock.no4sU.A      Mock   0.0         A         no4sU   True\n",
      "Mock.1h.A        Mock.1h.A      Mock   1.0         A            1h  False\n",
      "Mock.2h.A        Mock.2h.A      Mock   2.0         A            2h  False\n",
      "Mock.2h.B        Mock.2h.B      Mock   2.0         B            2h  False\n",
      "Mock.3h.A        Mock.3h.A      Mock   3.0         A            3h  False\n",
      "Mock.4h.A        Mock.4h.A      Mock   4.0         A            4h  False\n",
      "SARS.no4sU.A  SARS.no4sU.A      SARS   0.0         A         no4sU   True\n",
      "SARS.1h.A        SARS.1h.A      SARS   1.0         A            1h  False\n",
      "SARS.2h.A        SARS.2h.A      SARS   2.0         A            2h  False\n",
      "SARS.2h.B        SARS.2h.B      SARS   2.0         B            2h  False\n",
      "SARS.3h.A        SARS.3h.A      SARS   3.0         A            3h  False\n",
      "SARS.4h.A        SARS.4h.A      SARS   4.0         A            4h  False\n"
     ]
    }
   ],
   "execution_count": 30
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "<span style=\"color:red\">There are also some build-in grandPy functions that add metadata,such as `compute_expression_percentage`: </span>",
   "id": "101fe9dfd05d16fc"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:09.424751Z",
     "start_time": "2025-06-22T13:51:09.422151Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# fehlt in processing.py\n",
    "# sars <- ComputeExpressionPercentage(sars,name = \"viral_percentage\",\n",
    "#                                     genes = GeneInfo(sars,\"Type\")==\"viral\")\n",
    "# ggplot(Coldata(sars),aes(Name,viral_percentage))+\n",
    "#   geom_bar(stat=\"identity\")+\n",
    "#   RotatateAxisLabels()+\n",
    "#   xlab(NULL)"
   ],
   "id": "db2f426f8c9178b",
   "outputs": [],
   "execution_count": 31
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Interestingly the 4sU-naive sample shows more viral gene expression, suggesting that 4sU had an effect on viral gene expression.\n",
    "\n",
    "Since this is such an important control, there is also a specialized plotting built into grandR for that:"
   ],
   "id": "70c0ec913a4ecc67"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:10.266308Z",
     "start_time": "2025-06-22T13:51:10.264335Z"
    }
   },
   "cell_type": "code",
   "source": "# plot_type_distribution fehlt noch in plot.py",
   "id": "2b167882849620c9",
   "outputs": [],
   "execution_count": 32
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "There is a column in the `coldata` metadata table that has a special meaning: `Condition`. It is used by many functions as a default, e.g. to plot colors in the PCA or to model kinetics per conditions. It can be accessed by its own function:",
   "id": "e1c0d287dba91c85"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:10.907904Z",
     "start_time": "2025-06-22T13:51:10.904697Z"
    }
   },
   "cell_type": "code",
   "source": "sars.condition # Levels ausgabe?",
   "id": "152e5ec83ef9567d",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Mock',\n",
       " 'Mock',\n",
       " 'Mock',\n",
       " 'Mock',\n",
       " 'Mock',\n",
       " 'Mock',\n",
       " 'SARS',\n",
       " 'SARS',\n",
       " 'SARS',\n",
       " 'SARS',\n",
       " 'SARS',\n",
       " 'SARS']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 33
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "and it can be set either directly:",
   "id": "99689e2100e215a6"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:29.762839Z",
     "start_time": "2025-06-22T13:51:29.760760Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# new = sars.coldata[\"saved\"] = sars.coldata[\"Condition\"]\n",
    "# new.coldata[\"Condition\"] = [\"control\"] * 6 + [\"infected\"] * 6\n",
    "# print(sars.coldata)\n",
    "\n",
    "# es fehlt der vorausbau durch den aufruf von compute_expression_percentage\n",
    "# auch noch falsch, da immutability uns keine Spalten ergänzen lässt"
   ],
   "id": "b857aa2c79da50ef",
   "outputs": [],
   "execution_count": 37
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "or from one or several columns of the metadata (here this is not really reasonable, but there are situations where combining more than one metadata column makes sense):",
   "id": "10824f932fb20b4a"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-22T13:51:30.946273Z",
     "start_time": "2025-06-22T13:51:30.944278Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Condition(sars)<-c(\"saved\",\"Replicate\")   # set it by combining to other columns from the Coldata\n",
    "# Condition(sars)\n",
    "\n",
    "# Condition(sars)<-\"saved\"                  # set it to one other column from the Coldata\n",
    "# Condition(sars)"
   ],
   "id": "1319a946dab6d6d5",
   "outputs": [],
   "execution_count": 38
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
