{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Loading data and working with grandPy objects\n",
    "\n",
    "\n",
    "GrandPy is a Python package for the analysis of RNA-seq experiments involving metabolic RNA labeling with nucleotide conversion, such as SLAM-seq experiments [[1]](https://www.nature.com/articles/nmeth.4435). In such experiments, nucleoside analogs such as 4sU are added to living cells, which take it up and incorporate it into newly synthesized RNA. Before sequencing, 4sU is converted into a cytosin analog. Reads covering 4sU sites therefore have characteristic T-to-C mismatches after read mapping, in principle providing the opportunity to differentiate newly synthesized (during the time of labeling) from preexisting RNA.\n",
    "\n",
    "Confounders such as sequencing errors or read that originate from newly synthesized RNA but, by chance, do not cover sites of 4sU incorporation (usually 20-80% of all \"new read\") can be handled using specialized methods such as GRAND-SLAM [[2]](https://academic.oup.com/bioinformatics/article/34/13/i218/5045735?login=true).\n",
    "\n",
    "# Reading in the data\n",
    "\n",
    "Throughout this vignette, we will be using the GRAND-SLAM processed SLAM-seq data set from Finkel et al.2021 [[3]](https://www.nature.com/articles/s41586-021-03610-3). The data set contains time series (progressive labeling) samples from a human epithelial cell line (Calu3 cells); half of the samples were infected with SARS-CoV-2 for different periods of time.\n",
    "\n",
    "The output of GRAND-SLAM is a tsv file where rows are genes and columns are read counts and other statistics (e.g., the new-to-total RNA ratio) for all samples. The data set is available on zenodo (\"https://zenodo.org/record/5834034/files/sars.tsv.gz\"). We start by reading this file into Python:"
   ],
   "id": "999882dcca517e57"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Package installation\n",
    "from Py.load import *\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "\n",
    "sars = read_grand(\"https://zenodo.org/record/5834034/files/sars.tsv.gz\", design=(\"Condition\", \"dur.4sU\", \"Replicate\"))\n",
    "print(sars.columns)"
   ],
   "id": "829e05ae136792e5"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "When reading in the file, we have to define the `design` vector. This is used to infer metadata automatically from sample names. Here sample names consist of three parts separated by dots as shown above (the Columns function returns the sample names or cell ids when analyzing a single cell data set). Each part in the sample name represents an aspect of the design. For example, the sample named Mock.2h.A is a sample from the mock condition (i.e. not infected by SARS-CoV-2), subjected to metabolic labeling for 2 hours, and is the first replicate (i.e. replicate \"A\"). This sample name is consistent with the three element design vector used above. It is possible to specify other design elements (of course the samples would have to be named accordingly). A list of reasonable options is predefined in the dictionary `DESIGN_KEYS`.\n",
    "\n",
    "There are names (i.e. the things you specify in the design vector) that have additional semantics. For example, for the name `duration.4sU` the values are interpreted like this: 4h is converted into the number 4, 30min into 0.5, and no4sU into 0. For more information, see Paragraph *Column Metadata*). The design vector is mandatory. Attempting to read in the data without it results in an error:"
   ],
   "id": "f2c4e6d3dd877aa5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars_wrong = read_grand(\"https://zenodo.org/record/5834034/files/sars.tsv.gz\")",
   "id": "ea9457e66623a615"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Alternatively, a table containing the metadata can be specified. Make sure that it contains a `Name` column matching the names in the GRAND-SLAM output table:",
   "id": "9cf9341cc5fad6dc"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "import pandas as pd\n",
    "metadata = pd.DataFrame({\n",
    "    \"Name\": [\n",
    "        \"Mock.no4sU.A\",\"Mock.1h.A\",\"Mock.2h.A\",\"Mock.2h.B\",\n",
    "        \"Mock.3h.A\",\"Mock.4h.A\",\"SARS.no4sU.A\",\"SARS.1h.A\",\n",
    "        \"SARS.2h.A\",\"SARS.2h.B\",\"SARS.3h.A\",\"SARS.4h.A\"\n",
    "    ],\n",
    "    \"Condition\": [\"Mock\"] * 6 + [\"SARS\"] * 6\n",
    "})\n",
    "\n",
    "sars_meta = read_grand(\"../data/sars.tsv\", design=metadata)\n",
    "print(sars_meta)\n",
    "\n",
    "#TODO: Warning??"
   ],
   "id": "bcd5f3273e797a66"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# What is the grandPy object\n",
    "\n",
    "`read_grand` returns a grandPy object, which contains\n",
    "\n",
    "1. metadata for genes (gene_info)\n",
    "2. metadata for samples/cells (as inferred from the sample names by the design parameter) (coldata)\n",
    "3. all data matrices (counts, normalized counts, ntrs, etc. these types of data are called \"slots\") (slots)\n",
    "4. analysis results (analyses)\n",
    "\n",
    "Metadata (1. and 2.) are described below. How to work with the data matrices and analysis results is described in a separate [vignette](vignette_04_working_with_data_matrices_and_analysis_results.ipynb). <span style=\"color:red\">work in progress</span>\n",
    "\n",
    "# Working with grandPy objects\n",
    "\n",
    "Here we will see how to work with grandPy objects in general. A short summary can be displayed when `print`ing the object, and there are several functions to retrieve general information about the object:"
   ],
   "id": "bd2b70091bffe356"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars)",
   "id": "1e98b8e3f1ae9f69"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.title)",
   "id": "e566e8e67fba247b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(len(sars.genes))",
   "id": "df6513bc27e5a022"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(len(sars.columns))",
   "id": "ea72f54af634c024"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "\n",
    "It is straight-forward to filter genes:"
   ],
   "id": "3ce3f6e5ee092404"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.filter_genes()\n",
    "print(len(sars.genes))"
   ],
   "id": "b6248733969363b7"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "By default genes are retained if they have 100 read counts in at least half of the samples (or cells). There are many options how to filter by genes (note that `filter_genes` returns a new grandPy object, and below we directly call `len()` on this new object to check how many genes are retained by filtering):",
   "id": "7c1eb2b43974709e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(f\"Genes with at least 1000 read counts in half of the columns: {len((sars.filter_genes(min_expression=1000, return_genes=True)))}\")",
   "id": "50ade3444fe08070"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(f\"Genes with at least 1000 read counts in half of the columns (retain two genes that are otherwise filtered): {len((sars.filter_genes(min_expression=1000, keep=[\"ATF3\", \"ZC3H12A\"], return_genes=True)))}\")",
   "id": "ba4b996b0d0e1b50"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(f\"Keep only these two genes: {len((sars.filter_genes(use=[\"ATF3\", \"ZC3H12A\"],return_genes=True)))}\")",
   "id": "7e7c9572dcd0610"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.normalize_tpm()\n",
    "print(\"Genes with at least 10 TPM in half of the columns:\", len(sars.filter_genes(mode_slot=\"tpm\", min_expression=10, min_columns=sars.coldata.shape[0]/2, return_genes=True)))\n"
   ],
   "id": "cc38652e0cc8e66a"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "`filter_genes` essentially removes rows from the data slots. It is also possible to remove columns (i.e. samples or cells). See this example, it closely resembles pandas subsetting:",
   "id": "2b7b14c725832c58"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "mock = sars[:, sars.coldata[\"Condition\"] == \"Mock\"]\n",
    "print(mock)"
   ],
   "id": "f742582aa4fc54b0"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "<span style=\"color:red\">The new grandPy object now only has 6 columns. The `columns`parameter to subset must be a logical vector, and you can use the names of the column metadata table (see below) as variables (i.e. the parameter here is a logical vector with all samples being TRUE where the `Condition` column is equal to \"Mock\".\n",
    "A closely related function is `split`, which returns a list of several grandPy objects, each composed of samples having the same `Condition`.</span>"
   ],
   "id": "13cbca0ae6dc3637"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "object1, object2 = sars.split(by=\"Condition\")\n",
    "print(object1, \"\\n\", object2)"
   ],
   "id": "3ab51a0dfb2a5bbe"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(object1.columns, \"\\n\", object2.columns)",
   "id": "d2aec4eba01a6904"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "The inverse of `split` is `merge`:",
   "id": "9b633338e50d579a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars_mock = object1.merge(object2, axis=0)\n",
    "print(sars_mock.columns)"
   ],
   "id": "57eae5642a507347"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "`concat` can also be used, see below:",
   "id": "37177dad57efa879"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from Py.utils import concat\n",
    "sars_mock = concat([object2, object1], axis=0)\n",
    "print(sars_mock.columns)"
   ],
   "id": "8e52c12fdb6d508e"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Note that we merged such that now we have first the SARS samples and then the Mock samples. We can reorder by slightly abusing subsetting.",
   "id": "d220259dc63135fc"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars_mock = sars_mock[:, list(range(6,12)) + list(range(0,6))]\n",
    "print(sars_mock.columns)"
   ],
   "id": "f959c3ef69152575"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Gene metadata\n",
    "\n",
    "Here we see how to work with metadata for genes. The gene metadata essentially is a table that can be retrieved using the `gene_info` function:"
   ],
   "id": "5a3fffda3a62fedc"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.gene_info.head(10))",
   "id": "ad5cfc03f727807e"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Each gene has associated gene ids and symbols. Gene ids and symbols as well as the transcript length are part of GRAND-SLAM output. The `Type` column is inferred automatically (see below).\n",
    "\n",
    "Genes can be identified with the `genes` & `get_genes` functions:"
   ],
   "id": "584113918c6b52bf"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.genes[:20])",
   "id": "e8c6a8f7387fc79a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.get_genes(get_gene_symbols=False)[:20])",
   "id": "9a38a1f9cda439e4"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.get_genes(genes=[\"MYC\", \"ORF1ab\"],get_gene_symbols=False)[:20])",
   "id": "d6f7e012cba9a786"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.get_genes(genes=\"YC\", regex=True))",
   "id": "f2c63b7adfa7f107"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "During reading the data into Python using `read_grand`, the `Type` column is inferred using the `classify_genes` function. By default, this will recognize mitochondrial genes (MT prefix of the gene symbol), ERCC spike-ins, and Ensembl gene identifiers (which it will call \"cellular\"). Here we also have the viral genes, which are not properly recognized:",
   "id": "cab11efcd67510bc"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.gene_info[\"Type\"].value_counts())",
   "id": "c6a39765c768798f"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "If you want to define your own types, you can do this easily by specifying the `classify_genes` parameter when read in your data:",
   "id": "395603b58a5132f0"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "viral_genes = ['ORF3a','E','M','ORF6','ORF7a','ORF7b','ORF8','N','ORF10','ORF1ab','S']\n",
    "sars = read_grand(\"../data/sars.tsv\", design=(\"Condition\", \"dur.4sU\", \"Replicate\"), classification_genes=viral_genes, classification_genes_label=\"viral\")\n",
    "\n",
    "print(sars.gene_info[\"Type\"].value_counts())"
   ],
   "id": "2f8bab324c4ad67f"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Note that each parameter to `classify_genes` must be named (`viral`) and must be a function that takes the gene metadata table and returns a logical vector.\n",
    "\n",
    "The `classify_genes` function has one additional important parameter, which defines how \"Unknown\" types are supposed to be called. For this data set, a similar behavior as above can be accomplished by:"
   ],
   "id": "eb2365fc9ff3a973"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = read_grand(\"../data/sars.tsv\", design=(\"Condition\", \"dur.4sU\", \"Replicate\"), classify_genes_func=lambda df: classify_genes(df, name_unknown=\"viral\"))\n",
    "\n",
    "print(sars.gene_info[\"Type\"].value_counts())"
   ],
   "id": "2597f157d766cb97"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "It is also straight-forward to add additional gene metadata:",
   "id": "f6d4a94fcc21c23"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.with_gene_info(\"length.category\", pd.cut(\n",
    "    sars.gene_info[\"Length\"],\n",
    "    bins=[0, 2000, 5000, np.inf],\n",
    "    labels=[\"Short\", \"Medium\", \"Long\"]\n",
    "))\n",
    "\n",
    "print(sars.gene_info[\"length.category\"].value_counts())"
   ],
   "id": "a1b8768f2176cbc9"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Column Metadata\n",
    "\n",
    "Samples for bulk experiments and cells in single cell experiments are in grandR jointly called \"columns\". The metadata for columns is a table that describes the experimental design we specified when reading in data in grandPy. It can be accessed via the `coldata` function. We can also see that the duration of 4sU has been interpreted and converted to a numeric value (compare \"dur.4sU\" with \"dur.4sU.original\")."
   ],
   "id": "96dc48a6b79503b6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.coldata",
   "id": "e577016d4e366913"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Additional semantics can also be defined. This is done using the function `apply_design_semantics`, which creates a dictionary for the attribute `.attrs[“_semantics”]` in the `coldata` DataFrame. This dictionary is used to interpret design columns (e.g. time, concentration). The function `apply_design_semantics` is called automatically in `build_coldata()`.\n",
    "\n",
    "An important component here is that certain column contents (e.g. time specifications such as '5h' or '30min') can be converted into numerical values. There is already a predefined function `parse_time_string` in `grandPy` for this purpose:"
   ],
   "id": "a8328cb9acc33c9e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(semantics_time([\"5h\", \"30min\", \"no4sU\"], \"Test\"))",
   "id": "59ea25da13b8b348"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "We can easily define our own function like this:",
   "id": "b6bf4168944231d7"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "def my_semantics_time(values, name):\n",
    "    df = pd.DataFrame({name: values})\n",
    "    df[name] = df[name].map(parse_time_string)\n",
    "    h = df[name] + 3\n",
    "    def fmt(x):\n",
    "        return f\"{int(x) if x.is_integer() else x}hpi\"\n",
    "    df[\"hpi\"] = h.map(fmt)\n",
    "    df.index = range(1, len(df) + 1)\n",
    "    return df\n",
    "\n",
    "print(my_semantics_time([\"5h\", \"30min\", \"no4sU\"], \"Test\"))"
   ],
   "id": "df3b2cd24237bc93"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "<span style=\"color:red\"> Here, it is important to mention that at 3h post infection, 4sU was added to the cells for 1, 2, 3 or 4h. The two no4sU samples are also 3h post infection. This function can now be used as `semantics`parameter for `read_grand` like this:</span>",
   "id": "b52b0d6bc198ee6b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# # ???\n",
    "#\n",
    "# sars.meta <- ReadGRAND(system.file(\"extdata\", \"sars.tsv.gz\", package = \"grandR\"),\n",
    "#                    design=function(names) ??????\n",
    "#                      MakeColdata(names,\n",
    "#                                  c(\"Cell\",Design$dur.4sU,Design$Replicate),\n",
    "#                                  semantics=DesignSemantics(duration.4sU=my.semantics.time)\n",
    "#                                  ),\n",
    "#                  verbose=TRUE)\n",
    "\n",
    "sars_meta = read_grand(\"../data/sars_R.tsv\", design=(\"Cell\", \"dur.4sU\", \"Replicate\"))\n",
    "sars_meta.coldata"
   ],
   "id": "c55f28114a333340"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "As mentioned above, it is in most cases easier to add additional metadata after loading.The infection time point can also be added by:",
   "id": "8d610b563e115356"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "hpi_series = sars.coldata[\"duration.4sU\"].apply(lambda x: f\"{x + 3}hpi\")\n",
    "sars = sars.with_coldata(\"hpi\", hpi_series)\n",
    "sars.coldata"
   ],
   "id": "35dc1234809dfab9"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "There are also some build-in grandPy functions that add metadata,such as `compute_expression_percentage`:",
   "id": "5bc70dd39d9386c9"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from Py.processing import _compute_expression_percentage\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "sars = sars.compute_expression_percentage(name=\"viral_percentage\", genes=(sars.gene_info[\"Type\"]==\"viral\"))\n",
    "\n",
    "df = sars.coldata.copy()\n",
    "df[\"Sample\"] = df.index\n",
    "\n",
    "sns.set_style(\"whitegrid\")\n",
    "fig, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.barplot(data=df, x=\"Sample\", y=\"viral_percentage\", ax=ax)\n",
    "ax.set_axisbelow(True)\n",
    "ax.grid(axis=\"y\", linestyle=\"--\", alpha=0.7)\n",
    "ax.grid(axis=\"x\", linestyle=\"--\", alpha=0.7)\n",
    "ax.tick_params(axis=\"x\", rotation=90)\n",
    "plt.subplots_adjust(bottom=0.3)\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"viral_percentage (%)\")\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ],
   "id": "1c1591e82cc2400"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Interestingly the 4sU-naive sample shows more viral gene expression, suggesting that 4sU had an effect on viral gene expression.\n",
    "\n",
    "Since this is such an important control, there is also a specialized plotting built into grandR for that:"
   ],
   "id": "75103105eb1c326f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from Py.plot import *\n",
    "\n",
    "plot_type_distribution(sars, relative=True)\n",
    "\n",
    "# Bars nicht aufeinander?"
   ],
   "id": "d3d8230651e4ea38"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "There is a column in the `coldata` metadata table that has a special meaning: `Condition`. It is used by many functions as a default, e.g. to plot colors in the PCA or to model kinetics per conditions. It can be accessed by its own function:",
   "id": "5115a965b132bfe"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.condition)",
   "id": "ce76b65cd524afe8"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "and it can be set either directly:",
   "id": "fca7c23c953becea"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.with_coldata(\"saved\", sars.condition)\n",
    "new_conditions = [\"control\"] * 6 + [\"infected\"] * 6\n",
    "sars = sars.with_condition(new_conditions)\n",
    "sars.coldata"
   ],
   "id": "6e573b9ba525a77c"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "or from one or several columns of the metadata (here this is not really reasonable, but there are situations where combining more than one metadata column makes sense):",
   "id": "5271e7a63ad6e6de"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.with_condition([\"saved\", \"Replicate\"])\n",
    "print(sars.condition)"
   ],
   "id": "b8325445fbe8142d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.with_condition(\"saved\")\n",
    "print(sars.condition)\n"
   ],
   "id": "4cd215c36d22da01"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
