{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Loading data and working with grandPy objects\n",
    "\n",
    "<span style=\"color:green\">Notiz: alles was rot ist, ist noch nicht umgesetzt oder müsste noch überarbeitet werden durch Links etc.</span>\n",
    "\n",
    "GrandPy is a Python package for the analysis of RNA-seq experiments involving metabolic RNA labeling with nucleotide conversion, such as SLAM-seq experiments [[1]](https://www.nature.com/articles/nmeth.4435). In such experiments, nucleoside analogs such as 4sU are added to living cells, which take it up and incorporate it into newly synthesized RNA. Before sequencing, 4sU is converted into a cytosin analog. Reads covering 4sU sites therefore have characteristic T-to-C mismatches after read mapping, in principle providing the opportunity to differentiate newly synthesized (during the time of labeling) from preexisting RNA.\n",
    "\n",
    "Confounders such as sequencing errors or read that originate from newly synthesized RNA but, by chance, do not cover sites of 4sU incorporation (usually 20-80% of all \"new read\") can be handled using specialized methods such as GRAND-SLAM [[2]](https://academic.oup.com/bioinformatics/article/34/13/i218/5045735?login=true).\n",
    "\n",
    "# Reading in the data\n",
    "\n",
    "Throughout this vignette, we will be using the GRAND-SLAM processed SLAM-seq data set from Finkel et al.2021 [[3]](https://www.nature.com/articles/s41586-021-03610-3). The data set contains time series (progressive labeling) samples from a human epithelial cell line (Calu3 cells); half of the samples were infected with SARS-CoV-2 for different periods of time.\n",
    "\n",
    "The output of GRAND-SLAM is a tsv file where rows are genes and columns are read counts and other statistics (e.g., the new-to-total RNA ratio) for all samples. The data set is available on zenodo (\"https://zenodo.org/record/5834034/files/sars.tsv.gz\"). We start by reading this file into Python:"
   ],
   "id": "52df091142cc2d28"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:23.579735Z",
     "start_time": "2025-06-25T17:57:21.857407Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Package installation\n",
    "from Py.load import *\n",
    "sars = read_grand(\"https://zenodo.org/record/5834034/files/sars.tsv.gz\", design=(\"Condition\", \"Time\", \"Replicate\"))\n",
    "print(sars.columns)"
   ],
   "id": "ce3f00b85976da6",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected URL -> downloading to temp file\n",
      "Detected dense format -> using dense reader\n",
      "['Mock.no4sU.A', 'Mock.1h.A', 'Mock.2h.A', 'Mock.2h.B', 'Mock.3h.A', 'Mock.4h.A', 'SARS.no4sU.A', 'SARS.1h.A', 'SARS.2h.A', 'SARS.2h.B', 'SARS.3h.A', 'SARS.4h.A']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\PycharmProjects\\grandpy\\Py\\utils.py:61: UserWarning: 17 Duplicate gene symbols found: STPG4, PDE11A, TXNRD3NB, ARL14EPL, HIST1H3D (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\n",
      "  warnings.warn(f\"{len(duplicates_list)} Duplicate gene symbols found: {', '.join(duplicates_list[:5])} (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\")\n"
     ]
    }
   ],
   "execution_count": 1
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "When reading in the file, we have to define the `design` vector. This is used to infer metadata automatically from sample names. Here sample names consist of three parts separated by dots as shown above (the Columns function returns the sample names or cell ids when analyzing a single cell data set). Each part in the sample name represents an aspect of the design. For example, the sample named Mock.2h.A is a sample from the mock condition (i.e. not infected by SARS-CoV-2), subjected to metabolic labeling for 2 hours, and is the first replicate (i.e. replicate \"A\"). This sample name is consistent with the three element design vector used above. It is possible to specify other design elements (of course the samples would have to be named accordingly). A list of reasonable options is predefined in the dictionary `DESIGN_KEYS`.\n",
    "\n",
    "There are names (i.e. the things you specify in the design vector) that have additional semantics. For example, for the name `duration.4sU` the values are interpreted like this: 4h is converted into the number 4, 30min into 0.5, and no4sU into 0. For more information, see [below](#column-metadata)). The design vector is mandatory. Attempting to read in the data without it results in an error:\n",
    "\n",
    "<span style=\"color:red\">Notiz: \"Below\" funktioniert noch nicht\"</span>"
   ],
   "id": "70dc96beedb2eb02"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.541370Z",
     "start_time": "2025-06-25T17:57:23.583140Z"
    }
   },
   "cell_type": "code",
   "source": "sars_wrong = read_grand(\"https://zenodo.org/record/5834034/files/sars.tsv.gz\")",
   "id": "d9d464b16055a411",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected URL -> downloading to temp file\n",
      "Detected dense format -> using dense reader\n",
      "ValueError: Design must be specified.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\PycharmProjects\\grandpy\\Py\\utils.py:61: UserWarning: 17 Duplicate gene symbols found: STPG4, PDE11A, TXNRD3NB, ARL14EPL, HIST1H3D (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\n",
      "  warnings.warn(f\"{len(duplicates_list)} Duplicate gene symbols found: {', '.join(duplicates_list[:5])} (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\")\n"
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Alternatively, a table containing the metadata can be specified. Make sure that it contains a `Name` column matching the names in the GRAND-SLAM output table:",
   "id": "c301094d233f0520"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.771902Z",
     "start_time": "2025-06-25T17:57:24.640229Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "name = [\"Mock.no4sU.A\", \"Mock.1h.A\", \"Mock.2h.A\", \"Mock.2h.B\", \"Mock.3h.A\", \"Mock.4h.A\", \"SARS.no4sU.A\", \"SARS.1h.A\", \"SARS.2h.A\", \"SARS.2h.B\", \"SARS.3h.A\", \"SARS.4h.A\"]\n",
    "\n",
    "conditions = [\"Mock\"] * 6 + [\"SARS\"] * 6\n",
    "\n",
    "design_df = pd.DataFrame({\"Name\": name,\n",
    "                          \"Condition\": conditions})\n",
    "\n",
    "sars_meta = read_grand(\"../data/sars.tsv\", design = design_df)"
   ],
   "id": "d5f39c86bd40fb1d",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected dense format -> using dense reader\n",
      "Unexpected error: 'tuple' object has no attribute 'lower'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\PycharmProjects\\grandpy\\Py\\utils.py:61: UserWarning: 17 Duplicate gene symbols found: STPG4, PDE11A, TXNRD3NB, ARL14EPL, HIST1H3D (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\n",
      "  warnings.warn(f\"{len(duplicates_list)} Duplicate gene symbols found: {', '.join(duplicates_list[:5])} (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\")\n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# What is the grandPy object\n",
    "\n",
    "`read_grand` returns a grandPy object, which contains\n",
    "\n",
    "1. metadata for genes\n",
    "2. metadata for samples/cells (as inferred from the sample names by the design parameter)\n",
    "3. all data matrices (counts, normalized counts, ntrs, etc. these types of data are called \"slots\")\n",
    "4. analysis results\n",
    "\n",
    "Metadata (1. and 2.) are described below. How to work with the data matrices and analysis results is described in a separate [vignette](vignette_04_working_with_data_matrices_and_analysis_results.ipynb).\n",
    "\n",
    "# Working with grandPy objects\n",
    "\n",
    "Here we will see how to work with grandPy objects in general. A short summary can be displayed when `print`ing the object, and there are several functions to retrieve general information about the object:"
   ],
   "id": "98c2b82f1454b31a"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.780082Z",
     "start_time": "2025-06-25T17:57:24.777936Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(sars)\n",
    "# available analyses ???"
   ],
   "id": "e785e55dece18eb5",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GrandPy:\n",
      "Read from sars.tsv\n",
      "19659 genes, 12 samples/cells\n",
      "Available data slots: ['count', 'ntr', 'alpha', 'beta']\n",
      "Available analyses: []\n",
      "Available plots: {}\n",
      "Default data slot: count\n",
      "\n"
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.799091Z",
     "start_time": "2025-06-25T17:57:24.796957Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.title)",
   "id": "2af0cf929002159f",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sars.tsv\n"
     ]
    }
   ],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.827520Z",
     "start_time": "2025-06-25T17:57:24.824583Z"
    }
   },
   "cell_type": "code",
   "source": "print(len(sars.genes)) # so?",
   "id": "177ab54cb2c5be4b",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19659\n"
     ]
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.844358Z",
     "start_time": "2025-06-25T17:57:24.842103Z"
    }
   },
   "cell_type": "code",
   "source": "print(len(sars.coldata)) # so?",
   "id": "443d3ee445a9d8a2",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n"
     ]
    }
   ],
   "execution_count": 7
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "\n",
    "It is straight-forward to filter genes:"
   ],
   "id": "6a5077d14432faf4"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.888473Z",
     "start_time": "2025-06-25T17:57:24.858099Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Anpassen\n",
    "from Py.processing import filter_genes\n",
    "sars = filter_genes(sars)\n",
    "print(len(sars.genes))"
   ],
   "id": "43efc526cb2da73a",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9162\n"
     ]
    }
   ],
   "execution_count": 8
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "By default genes are retained if they have 100 read counts in at least half of the samples (or cells). There are many options how to filter by genes (note that `filter_genes` returns a new grandPy object, and below we directly call `len()` on this new object to check how many genes are retained by filtering):",
   "id": "e2bc88fb3b2d55bf"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.909322Z",
     "start_time": "2025-06-25T17:57:24.898579Z"
    }
   },
   "cell_type": "code",
   "source": "print(f\"Genes with at least 1000 read counts in half of the columns: {len((filter_genes(sars, min_expression=1000, return_genes=True)))}\")",
   "id": "207558afe7ae2c",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes with at least 1000 read counts in half of the columns: 1528\n"
     ]
    }
   ],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.928607Z",
     "start_time": "2025-06-25T17:57:24.913364Z"
    }
   },
   "cell_type": "code",
   "source": "print(f\"Genes with at least 1000 read counts in half of the columns (retain two genes that are otherwise filtered): {len((filter_genes(sars, min_expression=1000, keep=[\"ATF3\", \"ZC3H12A\"], return_genes=True)))}\")",
   "id": "87917c65d14ac4af",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes with at least 1000 read counts in half of the columns (retain two genes that are otherwise filtered): 1530\n"
     ]
    }
   ],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.942301Z",
     "start_time": "2025-06-25T17:57:24.934682Z"
    }
   },
   "cell_type": "code",
   "source": "print(f\"Keep only these two genes: {len((filter_genes(sars, use=[\"ATF3\", \"ZC3H12A\"],return_genes=True)))}\")",
   "id": "df519f5c6ddbd443",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Keep only these two genes: 2\n"
     ]
    }
   ],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.956439Z",
     "start_time": "2025-06-25T17:57:24.954006Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# R-code\n",
    "# sars <- NormalizeTPM(sars)     # compute transcript per million\n",
    "# cat(sprintf(\"Genes with at least 10 TPM in half of the columns: %d\\n\",\n",
    "#             nrow(FilterGenes(sars,mode.slot=\"tpm\",minval=10))))\n",
    "\n",
    "# fehlt bei processing.py\n",
    "\n",
    "# print(\"Genes with at least 10 TPM in half of the columns: \\n\", ...)"
   ],
   "id": "28b9ed5f500c97a0",
   "outputs": [],
   "execution_count": 12
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "`filter_genes` essentially removes rows from the data slots. It is also possible to remove columns (i.e. samples or cells). This is done using the subset function:",
   "id": "44fb34726a8bf5f2"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.985788Z",
     "start_time": "2025-06-25T17:57:24.979772Z"
    }
   },
   "cell_type": "code",
   "source": [
    "mock = sars[:, sars.coldata[\"Condition\"] == \"Mock\"]\n",
    "print(mock)\n",
    "# Analyses fehlerhaft\n",
    "# slot tpm fehlt\n",
    "# default data slot tpm fehlt"
   ],
   "id": "6256e056cf871101",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GrandPy:\n",
      "Read from sars.tsv\n",
      "9162 genes, 6 samples/cells\n",
      "Available data slots: ['count', 'ntr', 'alpha', 'beta']\n",
      "Available analyses: []\n",
      "Available plots: {}\n",
      "Default data slot: count\n",
      "\n"
     ]
    }
   ],
   "execution_count": 13
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "The new grandPy object now only has 6 columns. The `columns`parameter to subset must be a logical vector, and you can use the names of the column metadata table (see below) as variables (i.e. the parameter here is a logical vector with all samples being TRUE where the `Condition` column is equal to \"Mock\".\n",
    "\n",
    "<span style=\"color:red\">A closely related function is `split`, which returns a list of several grandPy objects, each composed of samples having the same `Condition`.</span>"
   ],
   "id": "aee49736178028a0"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:24.991811Z",
     "start_time": "2025-06-25T17:57:24.990036Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# split.list <- split(sars) # R-code\n",
    "# split.list\n",
    "\n",
    "# fehlt in grandPy.py"
   ],
   "id": "fd9fe0c617da067",
   "outputs": [],
   "execution_count": 14
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.009343Z",
     "start_time": "2025-06-25T17:57:25.007641Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# lapply(split.list,Columns) # R-code\n",
    "\n",
    "# lapply - in python maybe so?:"
   ],
   "id": "d5fa00606cbc923a",
   "outputs": [],
   "execution_count": 15
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "The inverse of `split` is `merge` <span>(concat?)</span>:",
   "id": "9763b2d6f968cff"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.028391Z",
     "start_time": "2025-06-25T17:57:25.025936Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# sars.mock <- merge(split.list$SARS,split.list$Mock)\n",
    "# Columns(sars.mock)\n",
    "\n",
    "# warte auf split und lapply"
   ],
   "id": "499c6fd9fb55d24c",
   "outputs": [],
   "execution_count": 16
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Note that we merged such that now we have first the SARS samples and then the Mock samples. We can reorder by slightly abusing `subset` (note that we actually do not omit any columns, but just define a different order):",
   "id": "b24aefda9fa945f0"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.041389Z",
     "start_time": "2025-06-25T17:57:25.039139Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# mock.sars <- subset(sars.mock,columns = c(7:12,1:6)) # R-code\n",
    "# Columns(mock.sars)\n",
    "\n",
    "# warte auf subset funktion"
   ],
   "id": "7ef18bdf7b67554",
   "outputs": [],
   "execution_count": 17
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Gene metadata\n",
    "\n",
    "Here we see how to work with metadata for genes. The gene metadata essentially is a table that can be retrieved using the `gene_info` function:"
   ],
   "id": "2e01839cc7dc2e97"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.055202Z",
     "start_time": "2025-06-25T17:57:25.049192Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.gene_info.head(10))",
   "id": "23ba05bb384a88fd",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          Symbol             Gene  Length      Type\n",
      "Symbol                                             \n",
      "MIB2        MIB2  ENSG00000197530    4247  Cellular\n",
      "OSBPL9    OSBPL9  ENSG00000117859    4520  Cellular\n",
      "BTF3L4    BTF3L4  ENSG00000134717    4703  Cellular\n",
      "ZFYVE9    ZFYVE9  ENSG00000157077    5194  Cellular\n",
      "PRPF38A  PRPF38A  ENSG00000134748    5274  Cellular\n",
      "AHCYL1    AHCYL1  ENSG00000168710    4313  Cellular\n",
      "ADGRL2    ADGRL2  ENSG00000117114    6302  Cellular\n",
      "HOOK1      HOOK1  ENSG00000134709    5857  Cellular\n",
      "NFIA        NFIA  ENSG00000162599    9487  Cellular\n",
      "PATJ        PATJ  ENSG00000132849    8505  Cellular\n"
     ]
    }
   ],
   "execution_count": 18
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Each gene has associated gene ids and symbols. Gene ids and symbols as well as the transcript length are part of GRAND-SLAM output. The `TRype` column is inferred automatically (see below).\n",
    "\n",
    "Genes can be identified with the `genes` & `get_genes` functions:"
   ],
   "id": "bef10b5acba3fe50"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.095535Z",
     "start_time": "2025-06-25T17:57:25.092620Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.genes[:20])",
   "id": "138a367610eac4cb",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MIB2', 'OSBPL9', 'BTF3L4', 'ZFYVE9', 'PRPF38A', 'AHCYL1', 'ADGRL2', 'HOOK1', 'NFIA', 'PATJ', 'VANGL1', 'GPR89B', 'GPSM2', 'ARHGEF10L', 'TMEM50A', 'TMEM57', 'ZNF593', 'RNF207', 'ZNF362', 'STX12']\n"
     ]
    }
   ],
   "execution_count": 19
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.115435Z",
     "start_time": "2025-06-25T17:57:25.112287Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.get_genes(use_gene_symbols=False)[:20])",
   "id": "e7bf64ff4ca1346d",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSG00000197530', 'ENSG00000117859', 'ENSG00000134717', 'ENSG00000157077', 'ENSG00000134748', 'ENSG00000168710', 'ENSG00000117114', 'ENSG00000134709', 'ENSG00000162599', 'ENSG00000132849', 'ENSG00000173218', 'ENSG00000188092', 'ENSG00000121957', 'ENSG00000074964', 'ENSG00000183726', 'ENSG00000204178', 'ENSG00000142684', 'ENSG00000158286', 'ENSG00000160094', 'ENSG00000117758']\n"
     ]
    }
   ],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.145871Z",
     "start_time": "2025-06-25T17:57:25.137492Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.get_genes(genes=[\"MYC\", \"ORF1ab\"],use_gene_symbols=False)[:20])",
   "id": "ac6fceebcaa7bbc0",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSG00000136997', 'ORF1ab']\n"
     ]
    }
   ],
   "execution_count": 21
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.166818Z",
     "start_time": "2025-06-25T17:57:25.159164Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.get_genes(genes=\"YC\", regex=True))",
   "id": "6daef7fe34ebb908",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NFYC', 'MYCBP', 'PYCR2', 'GLYCTK', 'FYCO1', 'CYCS', 'CYC1', 'MYC', 'PYCR3', 'MYCBP2', 'MLYCD', 'PYCR1', 'SYCP2']\n"
     ]
    }
   ],
   "execution_count": 22
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "During reading the data into Python using `read_grand`, the `Type` column is inferred using the `classify_genes` function. By default, this will recognize mitochondrial genes (MT prefix of the gene symbol), ERCC spike-ins, and Ensembl gene identifiers (which it will call \"cellular\"). Here we also have the viral genes, which are not properly recognized:",
   "id": "b38342b0a6354011"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.183768Z",
     "start_time": "2025-06-25T17:57:25.179583Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.gene_info[\"Type\"].value_counts()) # Table?",
   "id": "af69e3dd42aa747e",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Type\n",
      "Cellular    9151\n",
      "Unknown       11\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "execution_count": 23
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "If you want to define your own types, you can do this easily by specifying the `classify_genes` parameter when read in your data:",
   "id": "aa2cb42b2d41ab19"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.358857Z",
     "start_time": "2025-06-25T17:57:25.218660Z"
    }
   },
   "cell_type": "code",
   "source": [
    "viral_genes = ['ORF3a','E','M','ORF6','ORF7a','ORF7b','ORF8','N','ORF10','ORF1ab','S']\n",
    "sars = read_grand(\"../data/sars.tsv\", design=(\"Condition\", \"Time\", \"Replicate\"), classification_genes=viral_genes)\n",
    "\n",
    "print(sars.gene_info[\"Type\"].value_counts())"
   ],
   "id": "f4b3c6723f5d65bf",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected dense format -> using dense reader\n",
      "Type\n",
      "Cellular    19648\n",
      "Viral          11\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\PycharmProjects\\grandpy\\Py\\utils.py:61: UserWarning: 17 Duplicate gene symbols found: STPG4, PDE11A, TXNRD3NB, ARL14EPL, HIST1H3D (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\n",
      "  warnings.warn(f\"{len(duplicates_list)} Duplicate gene symbols found: {', '.join(duplicates_list[:5])} (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\")\n"
     ]
    }
   ],
   "execution_count": 24
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Note that each parameter to `classify_genes` must be named (`viral`) and must be a function that takes the gene metadata table and returns a logical vector.\n",
    "\n",
    "<span style=\"color:red\">The `classify_genes` function has one additional important parameter, which defines how \"Unknown\" types are supposed to be called. For this data set, a similar behavior as aboe can be accomplished by:</span>"
   ],
   "id": "58fcae1f6d980195"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.503605Z",
     "start_time": "2025-06-25T17:57:25.368990Z"
    }
   },
   "cell_type": "code",
   "source": [
    "sars = read_grand(\"../data/sars.tsv\", design=(\"Condition\", \"Time\", \"Replicate\"), classification_genes=None, classification_genes_label=\"Unknown\", classify_genes_func=None)\n",
    "\n",
    "print(sars.gene_info[\"Type\"].value_counts())\n",
    "\n",
    "# das ist noch nicht ganz richtig hier"
   ],
   "id": "d709b89ec499d58f",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected dense format -> using dense reader\n",
      "Type\n",
      "Cellular    19648\n",
      "Unknown        11\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\PycharmProjects\\grandpy\\Py\\utils.py:61: UserWarning: 17 Duplicate gene symbols found: STPG4, PDE11A, TXNRD3NB, ARL14EPL, HIST1H3D (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\n",
      "  warnings.warn(f\"{len(duplicates_list)} Duplicate gene symbols found: {', '.join(duplicates_list[:5])} (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\")\n"
     ]
    }
   ],
   "execution_count": 25
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "It is also straight-forward to add additional gene metadata:",
   "id": "2ff3b76ee181388c"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.509584Z",
     "start_time": "2025-06-25T17:57:25.507526Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# GeneInfo(sars,\"length.category\") <- cut(GeneInfo(sars,\"Length\"),\n",
    "#                                         breaks=c(0,2000,5000,Inf),\n",
    "#                                         labels = c(\"Short\",\"Medium\",\"Long\"))\n",
    "# table(GeneInfo(sars,\"length.category\")) # R-code"
   ],
   "id": "7a6193bae5f61f2f",
   "outputs": [],
   "execution_count": 26
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# <a id= \"column-metadata\">Column Metadata</a>\n",
    "\n",
    "<span style=\"color:red\">Anker funktioniert noch nicht</span>\n",
    "\n",
    "Samples for bulk experiments and cells in single cell experiments are in grandR jointly called \"columns\". The metadata for columns is a table that describes the experimental design we specified when reading in data in grandPy. It can be accessed via the `coldata` function. We can also see that the duration of 4sU has been interpreted and converted to a numeric value (compare \"Time\" with \"Time.original\")."
   ],
   "id": "1fc882e0543553d8"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:57:25.531446Z",
     "start_time": "2025-06-25T17:57:25.527256Z"
    }
   },
   "cell_type": "code",
   "source": "print(sars.coldata)",
   "id": "927647d64112464a",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      Name Condition  Time Replicate Time.original  no4sU\n",
      "Name                                                                     \n",
      "Mock.no4sU.A  Mock.no4sU.A      Mock   0.0         A         no4sU   True\n",
      "Mock.1h.A        Mock.1h.A      Mock   1.0         A            1h  False\n",
      "Mock.2h.A        Mock.2h.A      Mock   2.0         A            2h  False\n",
      "Mock.2h.B        Mock.2h.B      Mock   2.0         B            2h  False\n",
      "Mock.3h.A        Mock.3h.A      Mock   3.0         A            3h  False\n",
      "Mock.4h.A        Mock.4h.A      Mock   4.0         A            4h  False\n",
      "SARS.no4sU.A  SARS.no4sU.A      SARS   0.0         A         no4sU   True\n",
      "SARS.1h.A        SARS.1h.A      SARS   1.0         A            1h  False\n",
      "SARS.2h.A        SARS.2h.A      SARS   2.0         A            2h  False\n",
      "SARS.2h.B        SARS.2h.B      SARS   2.0         B            2h  False\n",
      "SARS.3h.A        SARS.3h.A      SARS   3.0         A            3h  False\n",
      "SARS.4h.A        SARS.4h.A      SARS   4.0         A            4h  False\n"
     ]
    }
   ],
   "execution_count": 27
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Additional semantics can also be defined, which is accomplished via the function `apply_design_semantics`, that generates a list for the `semantics` parameter of the function `build_coldata`, which in turn is used to infer metadata from sample names. We briefly explain these mechanisms with an example, but it is important to mention that in most cases, the desired metadata can be added after reading the data, as shown further below.\n",
    "\n",
    "<span style=\"color:red\">First, it is important to have a function that takes two parameters (a specific column of the original column of the original column metadata table + the name of this column) and returns a dataframe that is then `cbin`ed with the original column metadata table. There is one such predefined function in grandPy, which parses labeling durations:</span>"
   ],
   "id": "9ef9949fb6cfa4bb"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:58:56.199490Z",
     "start_time": "2025-06-25T17:58:56.195862Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from pandas import Series\n",
    "Series([\"5h\", \"30min\", \"no4sU\"]).map(parse_time_string)"
   ],
   "id": "f880daa19f81a7c9",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    5.0\n",
       "1    0.5\n",
       "2    0.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 37
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "We can easily define our own function like this:",
   "id": "e44a29c42585f3e8"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:58:59.508516Z",
     "start_time": "2025-06-25T17:58:59.502736Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def my_semantics_time(values, name):\n",
    "    df = pd.DataFrame({name: values})\n",
    "    df[name + \"_hr\"] = df[name].map(parse_time_string)\n",
    "    df[\"hpi\"] = df[name + \"_hr\"].apply(lambda x: f\"{x + 3}hpi\")\n",
    "    df.attrs[\"_semantics\"] = {name: \"time\"}\n",
    "    return df\n",
    "\n",
    "my_semantics_time([\"5h\", \"30min\", \"no4sU\"], \"Test\")"
   ],
   "id": "caaa8d9710463465",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    Test  Test_hr     hpi\n",
       "0     5h      5.0  8.0hpi\n",
       "1  30min      0.5  3.5hpi\n",
       "2  no4sU      0.0  3.0hpi"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Test</th>\n",
       "      <th>Test_hr</th>\n",
       "      <th>hpi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5h</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0hpi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30min</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.5hpi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>no4sU</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0hpi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 38
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "<span style=\"color:red\"> Here, it is important to mention that at 3h post infection, 4sU was added to the cells for 1, 2, 3 or 4h. The two no4sU samples are also 3h post infection. This function can now be used as `semantics`parameter for `read_grand` like this:</span>",
   "id": "697021a3ed1bbf0f"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T17:59:22.438655Z",
     "start_time": "2025-06-25T17:59:22.436150Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# sars.meta <- ReadGRAND(system.file(\"extdata\", \"sars.tsv.gz\", package = \"grandR\"),\n",
    "#                    design=function(names)\n",
    "#                      MakeColdata(names,\n",
    "#                                  c(\"Cell\",Design$dur.4sU,Design$Replicate),\n",
    "#                                  semantics=DesignSemantics(duration.4sU=my.semantics.time)\n",
    "#                                  ),\n",
    "#                  verbose=TRUE)"
   ],
   "id": "a3f28343d9b9eaf9",
   "outputs": [],
   "execution_count": 39
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "As mentioned above, it is in most cases easier to add additional metadata after loading.The infection time point can also be added by:",
   "id": "8f0614f9c546031c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars.coldata[\"hpi\"] = sars.coldata[\"Time\"].apply(lambda x: f\"{x + 3}hpi\")\n",
    "print(sars.coldata) # da es immutable ist, kann ich keine Spalten anfügen ..."
   ],
   "id": "19da886a06c5e828"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "<span style=\"color:red\">There are also some build-in grandPy functions that add metadata,such as `compute_expression_percentage`: </span>",
   "id": "4c9c83869e3419ff"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T18:00:34.060459Z",
     "start_time": "2025-06-25T18:00:34.058383Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# fehlt in processing.py\n",
    "\n",
    "# sars <- ComputeExpressionPercentage(sars,name = \"viral_percentage\",\n",
    "#                                     genes = GeneInfo(sars,\"Type\")==\"viral\")\n",
    "# ggplot(Coldata(sars),aes(Name,viral_percentage))+\n",
    "#   geom_bar(stat=\"identity\")+\n",
    "#   RotatateAxisLabels()+\n",
    "#   xlab(NULL)\n",
    "\n",
    "# R-code"
   ],
   "id": "b890eb2718913448",
   "outputs": [],
   "execution_count": 42
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Interestingly the 4sU-naive sample shows more viral gene expression, suggesting that 4sU had an effect on viral gene expression.\n",
    "\n",
    "Since this is such an important control, there is also a specialized plotting built into grandR for that:"
   ],
   "id": "70c0ec913a4ecc67"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T18:00:34.967926Z",
     "start_time": "2025-06-25T18:00:34.965849Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# PlotTypeDistribution(sars,relative = TRUE)\n",
    "\n",
    "# plot_type_distribution fehlt noch in plot.py"
   ],
   "id": "6772d390400227f2",
   "outputs": [],
   "execution_count": 43
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "There is a column in the `coldata` metadata table that has a special meaning: `Condition`. It is used by many functions as a default, e.g. to plot colors in the PCA or to model kinetics per conditions. It can be accessed by its own function:",
   "id": "e1c0d287dba91c85"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T18:00:51.963635Z",
     "start_time": "2025-06-25T18:00:51.960412Z"
    }
   },
   "cell_type": "code",
   "source": "sars.condition # Levels ausgabe?",
   "id": "b8aabd8364225141",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Mock',\n",
       " 'Mock',\n",
       " 'Mock',\n",
       " 'Mock',\n",
       " 'Mock',\n",
       " 'Mock',\n",
       " 'SARS',\n",
       " 'SARS',\n",
       " 'SARS',\n",
       " 'SARS',\n",
       " 'SARS',\n",
       " 'SARS']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 44
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "and it can be set either directly:",
   "id": "7d793c3b438b348c"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-25T18:01:13.959472Z",
     "start_time": "2025-06-25T18:01:13.957434Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# new = sars.coldata[\"saved\"] = sars.coldata[\"Condition\"]\n",
    "# new.coldata[\"Condition\"] = [\"control\"] * 6 + [\"infected\"] * 6\n",
    "# print(sars.coldata)\n",
    "\n",
    "# es fehlt der vorausbau durch den aufruf von compute_expression_percentage\n",
    "# auch noch falsch, da immutability uns keine Spalten ergänzen lässt"
   ],
   "id": "7132527ab055231c",
   "outputs": [],
   "execution_count": 45
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "or from one or several columns of the metadata (here this is not really reasonable, but there are situations where combining more than one metadata column makes sense):",
   "id": "10824f932fb20b4a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Condition(sars)<-c(\"saved\",\"Replicate\")   # set it by combining to other columns from the Coldata\n",
    "# Condition(sars)\n",
    "\n",
    "# Condition(sars)<-\"saved\"                  # set it to one other column from the Coldata\n",
    "# Condition(sars)"
   ],
   "id": "1be52db45f9c9914"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
