{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Differential expression\n",
    "\n",
    "One of the main applications of metabolic RNA labeling is to identify differentially expressed genes upon some perturbation [[1]](https://www.nature.com/articles/s41586-019-1369-y). The main benefit of using metabolic RNA labeling is that short-term changes can be resolved much better as compared to standard RNA-seq, since any changes of transcription that indeed happened are masked to some extent by pre-existing RNA.\n",
    "\n",
    "To perform differential analyses for data from [[2]](https://www.nature.com/articles/s41586-021-03689-8). These are SLAM-seq data from multiple time points (1h, 2h, 4h, 6h and 20h) after acute depletion of BANP. BANP is a transcriptional activator that binds to unmethylated CGCG motifs in CpG island promoters.\n",
    "\n",
    "# Loading data and QC\n",
    "We first load the grandPy package and then read the GRAND-SLAM output table directly from zenodo:"
   ],
   "id": "e0b9d448903a9f6"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-26T15:46:41.253916Z",
     "start_time": "2025-06-26T15:46:39.600329Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from numpy.ma.core import min_val\n",
    "\n",
    "# loading the grandPy package anpassen\n",
    "from Py.load import *\n",
    "banp = read_grand(\"https://zenodo.org/record/6976391/files/BANP.tsv.gz\", design=(\"Cell\", \"Experimental.time\", \"Genotype\", \"dur.4sU\", \"has4.U\", \"Replicate\"))"
   ],
   "id": "265c473a229140f9",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected URL -> downloading to temp file\n",
      "Detected dense format -> using dense reader\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\PycharmProjects\\grandpy\\Py\\utils.py:61: UserWarning: 34 Duplicate gene symbols found: Gm16701, Sept2, Gm2464, Jakmip1, Dancr (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\n",
      "  warnings.warn(f\"{len(duplicates_list)} Duplicate gene symbols found: {', '.join(duplicates_list[:5])} (first 5); they have been renamed to ensure uniqueness (e.g., MATR3 → MATR3_1).\")\n"
     ]
    }
   ],
   "execution_count": 19
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Refer to the [Loading data and working with grandPy objects](vignette_03_loading_data_and_working_with_grandPy_objects.ipynb) vignette to learn more about how to load data. Note that sample metadata has been automatically extracted from the sample names via the design parameter given to `read_grand`:",
   "id": "38dfa0b4bd2e7900"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-26T15:45:54.450663Z",
     "start_time": "2025-06-26T15:45:54.443929Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(banp.coldata)\n",
    "# duration.4sU.original?"
   ],
   "id": "87f858453760cdf9",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                  Name  Cell  \\\n",
      "Name                                                           \n",
      "mESC.0h.wt.0h.no4sU.A            mESC.0h.wt.0h.no4sU.A  mESC   \n",
      "mESC.0h.wt.0h.no4sU.B            mESC.0h.wt.0h.no4sU.B  mESC   \n",
      "mESC.0h.wt.0h.no4sU.C            mESC.0h.wt.0h.no4sU.C  mESC   \n",
      "mESC.0h.wt.2h.4sU.A                mESC.0h.wt.2h.4sU.A  mESC   \n",
      "mESC.0h.wt.2h.4sU.B                mESC.0h.wt.2h.4sU.B  mESC   \n",
      "mESC.0h.wt.2h.4sU.C                mESC.0h.wt.2h.4sU.C  mESC   \n",
      "mESC.0h.dTag.2h.4sU.A            mESC.0h.dTag.2h.4sU.A  mESC   \n",
      "mESC.0h.dTag.2h.4sU.B            mESC.0h.dTag.2h.4sU.B  mESC   \n",
      "mESC.1h.dTag.30min.4sU.A      mESC.1h.dTag.30min.4sU.A  mESC   \n",
      "mESC.1h.dTag.30min.4sU.B      mESC.1h.dTag.30min.4sU.B  mESC   \n",
      "mESC.1h.dTag.30min.4sU.C      mESC.1h.dTag.30min.4sU.C  mESC   \n",
      "mESC.2h.dTag.90min.4sU.A      mESC.2h.dTag.90min.4sU.A  mESC   \n",
      "mESC.2h.dTag.90min.4sU.B      mESC.2h.dTag.90min.4sU.B  mESC   \n",
      "mESC.2h.dTag.90min.4sU.C      mESC.2h.dTag.90min.4sU.C  mESC   \n",
      "mESC.4h.dTag.120min.4sU.A    mESC.4h.dTag.120min.4sU.A  mESC   \n",
      "mESC.4h.dTag.120min.4sU.B    mESC.4h.dTag.120min.4sU.B  mESC   \n",
      "mESC.4h.dTag.120min.4sU.C    mESC.4h.dTag.120min.4sU.C  mESC   \n",
      "mESC.6h.dTag.120min.4sU.A    mESC.6h.dTag.120min.4sU.A  mESC   \n",
      "mESC.6h.dTag.120min.4sU.B    mESC.6h.dTag.120min.4sU.B  mESC   \n",
      "mESC.6h.dTag.120min.4sU.C    mESC.6h.dTag.120min.4sU.C  mESC   \n",
      "mESC.20h.dTag.120min.4sU.A  mESC.20h.dTag.120min.4sU.A  mESC   \n",
      "mESC.20h.dTag.120min.4sU.B  mESC.20h.dTag.120min.4sU.B  mESC   \n",
      "mESC.20h.dTag.120min.4sU.C  mESC.20h.dTag.120min.4sU.C  mESC   \n",
      "\n",
      "                           Experimental.time Genotype dur.4sU has4.U  \\\n",
      "Name                                                                   \n",
      "mESC.0h.wt.0h.no4sU.A                     0h       wt      0h  no4sU   \n",
      "mESC.0h.wt.0h.no4sU.B                     0h       wt      0h  no4sU   \n",
      "mESC.0h.wt.0h.no4sU.C                     0h       wt      0h  no4sU   \n",
      "mESC.0h.wt.2h.4sU.A                       0h       wt      2h    4sU   \n",
      "mESC.0h.wt.2h.4sU.B                       0h       wt      2h    4sU   \n",
      "mESC.0h.wt.2h.4sU.C                       0h       wt      2h    4sU   \n",
      "mESC.0h.dTag.2h.4sU.A                     0h     dTag      2h    4sU   \n",
      "mESC.0h.dTag.2h.4sU.B                     0h     dTag      2h    4sU   \n",
      "mESC.1h.dTag.30min.4sU.A                  1h     dTag   30min    4sU   \n",
      "mESC.1h.dTag.30min.4sU.B                  1h     dTag   30min    4sU   \n",
      "mESC.1h.dTag.30min.4sU.C                  1h     dTag   30min    4sU   \n",
      "mESC.2h.dTag.90min.4sU.A                  2h     dTag   90min    4sU   \n",
      "mESC.2h.dTag.90min.4sU.B                  2h     dTag   90min    4sU   \n",
      "mESC.2h.dTag.90min.4sU.C                  2h     dTag   90min    4sU   \n",
      "mESC.4h.dTag.120min.4sU.A                 4h     dTag  120min    4sU   \n",
      "mESC.4h.dTag.120min.4sU.B                 4h     dTag  120min    4sU   \n",
      "mESC.4h.dTag.120min.4sU.C                 4h     dTag  120min    4sU   \n",
      "mESC.6h.dTag.120min.4sU.A                 6h     dTag  120min    4sU   \n",
      "mESC.6h.dTag.120min.4sU.B                 6h     dTag  120min    4sU   \n",
      "mESC.6h.dTag.120min.4sU.C                 6h     dTag  120min    4sU   \n",
      "mESC.20h.dTag.120min.4sU.A               20h     dTag  120min    4sU   \n",
      "mESC.20h.dTag.120min.4sU.B               20h     dTag  120min    4sU   \n",
      "mESC.20h.dTag.120min.4sU.C               20h     dTag  120min    4sU   \n",
      "\n",
      "                           Replicate  no4sU  \n",
      "Name                                         \n",
      "mESC.0h.wt.0h.no4sU.A              A  False  \n",
      "mESC.0h.wt.0h.no4sU.B              B  False  \n",
      "mESC.0h.wt.0h.no4sU.C              C  False  \n",
      "mESC.0h.wt.2h.4sU.A                A  False  \n",
      "mESC.0h.wt.2h.4sU.B                B  False  \n",
      "mESC.0h.wt.2h.4sU.C                C  False  \n",
      "mESC.0h.dTag.2h.4sU.A              A  False  \n",
      "mESC.0h.dTag.2h.4sU.B              B  False  \n",
      "mESC.1h.dTag.30min.4sU.A           A  False  \n",
      "mESC.1h.dTag.30min.4sU.B           B  False  \n",
      "mESC.1h.dTag.30min.4sU.C           C  False  \n",
      "mESC.2h.dTag.90min.4sU.A           A  False  \n",
      "mESC.2h.dTag.90min.4sU.B           B  False  \n",
      "mESC.2h.dTag.90min.4sU.C           C  False  \n",
      "mESC.4h.dTag.120min.4sU.A          A  False  \n",
      "mESC.4h.dTag.120min.4sU.B          B  False  \n",
      "mESC.4h.dTag.120min.4sU.C          C  False  \n",
      "mESC.6h.dTag.120min.4sU.A          A  False  \n",
      "mESC.6h.dTag.120min.4sU.B          B  False  \n",
      "mESC.6h.dTag.120min.4sU.C          C  False  \n",
      "mESC.20h.dTag.120min.4sU.A         A  False  \n",
      "mESC.20h.dTag.120min.4sU.B         B  False  \n",
      "mESC.20h.dTag.120min.4sU.C         C  False  \n"
     ]
    }
   ],
   "execution_count": 14
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "The experimental time column indicates the period of time the sample was treated with the dTAG13 compound that induces acute depletion of BANP within 30 minutes\n",
    "\n",
    "By default GRAND-SLAM will report data on all genes (with at least one mapped read), and `read_grand` will read all these genes from the output:"
   ],
   "id": "beb8461faa028991"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-26T15:46:05.291775Z",
     "start_time": "2025-06-26T15:46:05.289423Z"
    }
   },
   "cell_type": "code",
   "source": "print(banp)",
   "id": "6d61328a7e367fd5",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GrandPy:\n",
      "Read from BANP.tsv\n",
      "24616 genes, 23 samples/cells\n",
      "Available data slots: ['count', 'ntr', 'alpha', 'beta']\n",
      "Available analyses: []\n",
      "Available plots: {}\n",
      "Default data slot: count\n",
      "\n"
     ]
    }
   ],
   "execution_count": 15
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Thus, we filter to only include genes that have at least 100 reads in at least 11 samples:",
   "id": "3408eb905b639583"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-26T15:46:07.723961Z",
     "start_time": "2025-06-26T15:46:07.681616Z"
    }
   },
   "cell_type": "code",
   "source": [
    "banp = banp.filter_genes(min_expression=100, min_columns=11)\n",
    "print(banp)"
   ],
   "id": "7167108f7229ff33",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GrandPy:\n",
      "Read from BANP.tsv\n",
      "11147 genes, 23 samples/cells\n",
      "Available data slots: ['count', 'ntr', 'alpha', 'beta']\n",
      "Available analyses: []\n",
      "Available plots: {}\n",
      "Default data slot: count\n",
      "\n"
     ]
    }
   ],
   "execution_count": 16
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "The actual data is available in so-called \"data slots\". `read_grand` adds the read counts, new to toal RNA ratios (NTRs) and information on the NTR posterior distribution (alpha, beta).\n",
    "\n",
    "As a quick quality check, we can inspect a principal component analysis of all samples involved:"
   ],
   "id": "5020b693388cb345"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-27T12:28:39.498642Z",
     "start_time": "2025-06-27T12:28:38.158525Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from Py.plot import *\n",
    "# plot_pca(banp)"
   ],
   "id": "8d40e11efe5410f6",
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-26T15:35:35.712668Z",
     "start_time": "2025-06-26T15:35:35.711178Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "id": "6f6a2dbeee144bf2",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
