{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Working with data matrices and analysis results\n",
    "\n",
    "This notebook will show the most suitable commands to retrieve data from GrandPy object in different scenarios.\n",
    "\n",
    "Throughout this notebook, we will be using GRAND-SLAM processed SLAM-seq data set from [Finkel et al. (2021)](https://www.nature.com/articles/s41586-021-03610-3). The data set contains time series (progressive labeling) samples from a human epithelial cell line (Calu3 cells); half of the samples were infedctes with SARS-CoV-2 for different periods of time. For more on these initial commands see the [loading data notebook](../notebook_03_loading_data_and_working_with_grandpy_objects)."
   ],
   "id": "6af6edc69402a205"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "import warnings\n",
    "import pandas as pd\n",
    "from scripts.regsetup import description\n",
    "\n",
    "import grandpy as gp\n",
    "from grandpy import ModeSlot\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "\n",
    "sars = gp.read_grand(\"https://zenodo.org/record/5834034/files/sars.tsv.gz\", design=(\"Condition\", \"dur.4sU\", \"Replicate\"), classify_genes_func=lambda df: gp.classify_genes(df, cg_name=\"viral\"))"
   ],
   "id": "bf941953635d6990"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars = sars.filter_genes()",
   "id": "7a323e3761f2dfe1"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Data slots\n",
    "Data is organized in `slots`:"
   ],
   "id": "32ae83b6182efc8f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.slots",
   "id": "e477f9f626c30d80"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "To learn more about metadata, see the [loading data notebook](../notebook_03_loading_data_and_working_with_grandpy_objects). After loading GRAND-SLAM analysis results, the standard slots are \"count\" (read counts), \"ntr\" (the new-to-total RNA ratio), \"alpha\" and \"beta\" (the parameters for the Beta approximation of the NTR posterior distribution). Each of these slots contains a genes x columns matrix of numeric values. (columns are either samples or cells, depending on whether your data is bulk or single cell data)\n",
    "\n",
    "There is also a `default_slot`, which is used by many functions as a default parameter."
   ],
   "id": "2ba5ff6153f4f3f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.default_slot",
   "id": "1f108ba3e080e336"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "New slots are added by specific GrandPy methods such as `normalize()` or `normalize_tpm()`, which change the `default_slot`. It can also be set manually.",
   "id": "caf822f4f2d42d03"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.normalize()\n",
    "sars.default_slot"
   ],
   "id": "826755d22de6b9ff"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.with_default_slot(\"count\")\n",
    "sars.default_slot"
   ],
   "id": "b30bfcd8cd899c36"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.normalize_tpm(set_to_default = False)\n",
    "sars.default_slot"
   ],
   "id": "32c5026d223494d4"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.with_default_slot(\"norm\"))",
   "id": "d3090c52cf56ce6"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "There are also other GrandPy functions that add additional slots, but do not update the `default_slot` automatically:",
   "id": "a3957fef704a574d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.compute_ntr_ci()\n",
    "sars.default_slot"
   ],
   "id": "7299141eb9fb3029"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.slots",
   "id": "6406d3477159255c"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Analysis\n",
    "In addition to data slots, there is an additional kind of data that is part of a GrandPy object: analyses."
   ],
   "id": "ad07574284d5a124"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.analyses",
   "id": "97b0698a7d88efb"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "After loading data there are no analyses, but such data is added e.g. by performing modeling of progressive labeling time courses or analyzing gene expression (see the notebooks [kinetic modeling](../notebook_02_kinetic_modeling) and [differential expression](../notebook_01_differential_expression) for more on these)",
   "id": "7a9af80cb3665b48"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.fit_kinetics(steady_state = {\"Mock\": True, \"SARS\": False})\n",
    "sars = sars.compute_lfc(contrasts = sars.get_contrasts(contrast = [\"duration.4sU.original\", \"no4sU\"], group = \"Condition\", no4su = True))\n",
    "sars.analyses"
   ],
   "id": "ec96ba3fb3b1d3a4"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Both analysis methods, `fit_kinetics()` and `compute_lfc()` added multiple analyses: `fit_kinetics()` added an analysis for each `condition` whereas `compute_lfc()` added an analysis for each pairwise comparisons defined by `get_contrasts()` (see [differential expression](../notebook_01_differential_expression) for details).\n",
    "\n",
    "What is common to data slots and analyses is that both are tables with as many rows as there are genes. The main difference are their columns. In `slots` the columns correspond to the samples or cells (depending on whether data are bulk or single cell data). In `analyses` the columns are arbitrary, only depending on the kind of analysis performed.\n",
    "\n",
    "Analysis columns can be retrieved by setting the `description` to 'True' for `get_analyses()`:"
   ],
   "id": "72428a85e9669ca2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_analyses(description = True)",
   "id": "c1e69d21153ab309"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "We see that `fit_kinetics()` by default creates tables with two columns (*Synthesis* and *Half-life*) corresponding to the synthesis rate and RNA half-life for each gene. <span style=\"color:red;\">`compute_lfc()` creates a single column called *lfc* corresponding to the log2 fold changes for each gene.",
   "id": "a04ec9972075dcf3"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Retrieving data from slots or analyses\n",
    "There are three main methods you can use for retrieving the data from `slots`:\n",
    "- `get_table()`: Returns a DataFrame with genes as rows and columns made from potentially several slots\n",
    "- `get_data()`: Returns a DataFrame with the samples/cells as rows and slot data for particular genes in columns\n",
    "- `get_analysis_table()`: Returns a DataFrame with genes as rows and columns made from potentially several analyses\n",
    "\n",
    "# get_table\n",
    "Without any other parameters `get_table()` returns for all genes from `default_slot`:"
   ],
   "id": "a263cbcdc92323c2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(mode_slot = \"count\")",
   "id": "a903f5b13e1598b2"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "You can change the slot by specifying another `mode_slot` parameter:",
   "id": "3c10f0184d7f1cbb"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(mode_slot = [\"norm\", \"count\"]).columns",
   "id": "321f52ec10282b9"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "By using the `mode_slot` syntax (mode being either of `total`, `new` and `old`), you can also retrieve new RNA counts or new RNA normalized values:",
   "id": "b97b1dda88983261"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(mode_slot= gp.ModeSlot(\"new\", \"norm\"), ntr_nan = True)",
   "id": "a96a2c048bfae701"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Note that the no4sU columns only have NaN values. You can change this behaviour by specifying the `ntr_nan` parameter:",
   "id": "fe061d958572ffc9"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(mode_slot= gp.ModeSlot(\"new\", \"norm\"), ntr_nan = False)",
   "id": "ccc6b14800892a75"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "`get_analysis_table()` is used to retrieve the analysis results:",
   "id": "89ba06ef2adc5b73"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_analysis_table(\"kinetics\")",
   "id": "bb955e1f3661f810"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Note that you can also specify the full name (it actually is a regular expression that is matched against each analysis name).\n",
    "\n",
    "It is also easily possible to only retrieve data for specific columns (i.e., samples or cells) by using the `columns` parameter. Note that you can use names from the `coldata` table to construct a logical vector over the columns; using a character vector (to specify names) or a numeric vector (to specify positions) also works:"
   ],
   "id": "255c7a739bfb7b1f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(columns = (sars.coldata[\"duration.4sU\"] >= 2) & (sars.coldata[\"Condition\"] == \"Mock\"))",
   "id": "3f25becd2aeb5a1b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(columns = [\"Mock.no4sU.A\", \"SARS.no4sU.A\"])",
   "id": "b60915760f734de1"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(columns = sars.columns[3:6])",
   "id": "58917de873e46077"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "It is furthermore possible to only fetch data for specific genes, e.g. viral genes, using the `genes` parameter. It is either a logical vector, a numeric vector or a gene name/symol:",
   "id": "963a5212b5d97d5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(genes = (sars.gene_info[\"Type\"] == \"viral\"))",
   "id": "bc570dc1a92cbd1c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(genes = range(0,3))",
   "id": "447e083663de4a44"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(genes = \"MYC\")",
   "id": "a5bf7e8bbfbbb4e0"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Sometimes, it makes sense to add the `with_gene_info()` table (for more on gene metadata, see the [laoding data notebook](notebook_03_loading_data_and_working_with_grandPy_objects.ipynb)):",
   "id": "f3f1e4e00b4c5698"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "df = sars.get_table(mode_slot = \"norm\", with_gene_info = True)\n",
    "df"
   ],
   "id": "28915aac8ff5327b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "plt.figure(figsize=(6, 6))\n",
    "sns.scatterplot(\n",
    "    data=df,\n",
    "    x='SARS.4h.A',\n",
    "    y='SARS.no4sU.A',\n",
    "    hue='Type'\n",
    ")\n",
    "\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "\n",
    "lims = [\n",
    "    np.min([df['SARS.4h.A'].min(), df['SARS.no4sU.A'].min()]),\n",
    "    np.max([df['SARS.4h.A'].max(), df['SARS.no4sU.A'].max()]),\n",
    "]\n",
    "plt.plot(lims, lims, 'k-')\n",
    "plt.xlim(lims)\n",
    "plt.ylim(lims)\n",
    "\n",
    "plt.xlabel('SARS.4h.A')\n",
    "plt.ylabel('SARS.no4sU.A')\n",
    "plt.title('Log-Log Scatter Plot')\n",
    "plt.legend(title='Type')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ],
   "id": "c503a63283f4e787"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Finally, it is also straight-forward to get summarized values across samples or cells from the same `Condition`:",
   "id": "bc7846f100549a23"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(summarize = sars.get_summary_matrix())",
   "id": "5481242c83d611e7"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "This is accomplished by a \"summerize matrix\":",
   "id": "ca0ebe4f6db34196"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "smat = sars.get_summary_matrix()\n",
    "smat"
   ],
   "id": "efe6e4debf92ead4"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Instead of specifying `True` for the `summarize` parameter, you can also specify such a matrix:",
   "id": "8143b0187b44c059"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(summarize = smat)",
   "id": "c68f5204f6643a"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "For summarization, the summarize matrix is matrix-multiplied with the raw matrix. `get_summarized_matrix()` will generate a matrix with columns corresponding to `Condition` s:",
   "id": "eef5485e0a2e6d65"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.condition)",
   "id": "bb21aa9af01fbb35"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "By default, no4sU columns are removed (i.e. zero in the matrix), but the `no4sU` parameter can change this:",
   "id": "555cea7c9c6a2027"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_summary_matrix(no4su = True)",
   "id": "48824f5f8e6324fa"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "It is also possible to focus on specific columns (sampels or cells) only:",
   "id": "edef9103d97ddcae"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_summary_matrix(columns = sars.coldata[\"duration.4sU\"] < 4)",
   "id": "b72ff944932526a1"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "The default behaviour is to compute the average, this can be changed to computing sums:",
   "id": "b5e2fd42b6047c62"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_summary_matrix(average = False)",
   "id": "bb752273b97e8f12"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "As a final example, to get averaged normalized values for the 2h timepoint only:",
   "id": "b5a1e86747cfba9f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(summarize = sars.get_summary_matrix(columns = sars.coldata[\"duration.4sU\"] == 2))",
   "id": "7abaeef14492afec"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# get_data\n",
    "`get_data` is the \"little cousin\" of `get_table`: It returns a data frame with the samples or cells as rows and slot data, for either a singel gen or very few genes:"
   ],
   "id": "b8a16f0400bf21aa"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_data(genes = \"MYC\")",
   "id": "74f5ed865c301165"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Note that by default, the `Coldata` table is also added (for more column-metadata, see the [loading data notebook](notebook_03_loading_data_and_working_with_grandPy_objects.ipynb)). Note that in contrast to `get_table`, where you can add the `gene_info` table, i.e. gene metadta, here it is the columns metadata! This can be changed by using the `coldata` parameter:",
   "id": "6bf189d5e5ba273c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_data(genes = \"MYC\", with_coldata = False)",
   "id": "75a2633a4b5914d4"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "It is also possible to retrieve data for multiple genes and/or multiple slots and to restrict the columns:",
   "id": "c9035978accad16c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_data(genes = [\"MYC\", \"SRSF6\"], columns = sars.coldata[\"Condition\"] == \"Mock\", with_coldata = False)",
   "id": "dc19c25588098749"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_data(mode_slot = [(\"new_norm\"), (\"old_norm\")], genes = \"MYC\", columns = sars.coldata[\"Condition\"] == \"Mock\", with_coldata = False, ntr_nan = False)",
   "id": "a30d6271e6d5c1a5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_data(mode_slot = [(\"count\"), (\"norm\")], genes = [\"MYC\", \"SRSF6\"], columns = sars.coldata[\"Condition\"] == \"Mock\", with_coldata = False)",
   "id": "3228b9c84842968"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Finally, it is also possible to append multiple genes (and/or slots) not as columns, but as additional rows:",
   "id": "b67a29d01a0034df"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_data(genes = [\"MYC\", \"SRSF6\"], columns = sars.coldata[\"duration.4sU\"] < 2, by_rows = True)",
   "id": "6be29254385db56f"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "This can be quite helpful, as for the following example: We retrieve total, old new RNA for SRSF6 (only replicate A), and do this by rows. This way, the data can directly be used for mathplotlib/seaborn to plot the progressive labeling time course (note the much shorter half-life, which is the time where the new and old lines cross, for SARS as compared to Mock):",
   "id": "bf9b8517cd2ea6e4"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "df = sars.get_data(mode_slot = [\"old_norm\", \"new_norm\", \"total_norm\"], genes = \"SRSF6\",\n",
    "                   columns = sars.coldata[\"Replicate\"] == \"A\", by_rows = True)\n",
    "\n",
    "g = sns.FacetGrid(df, col= \"Condition\", col_wrap=4, height=4, sharey=True)\n",
    "\n",
    "g.map(sns.lineplot, 'duration.4sU', 'Value')\n",
    "\n",
    "g.add_legend()\n",
    "g.set_axis_labels(\"duration.4sU\", \"Value\")\n",
    "g.set_titles(col_template=\"{col_name}\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ],
   "id": "a3135cc974893f"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# get_analysis_table\n",
    "`get_analysis_table` "
   ],
   "id": "72a1ddd02d0b6d56"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
