{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Working with data matrices and analysis results\n",
    "\n",
    "This notebook will show the most suitable commands to retrieve data from grandPy object in different scenarios.\n",
    "\n",
    "Throughout this notebook, we will be using GRAND-SLAM processed SLAM-seq data set from Finkel et al. 2021 [[3]](https://www.nature.com/articles/s41586-021-03610-3). The data set contains time series (progressive labeling) samples from a human epithelial cell line (Calu3 cells); half of the samples were infedctes with SARS-CoV-2 for different periods of time. For more on these initial commands see the [Loading data](--) notebook."
   ],
   "id": "89e228aa8d71aded"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:09.910185Z",
     "start_time": "2025-07-29T11:43:08.354706Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import warnings\n",
    "import pandas as pd\n",
    "from scripts.regsetup import description\n",
    "\n",
    "import grandpy as gp\n",
    "from grandpy import ModeSlot\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "\n",
    "df = lambda x: x\n",
    "sars = gp.read_grand(\"https://zenodo.org/record/5834034/files/sars.tsv.gz\", design=(\"Condition\", \"dur.4sU\", \"Replicate\"), classify_genes_func=lambda df: gp.classify_genes(df, cg_name=\"viral\"))"
   ],
   "id": "db4490e451b55c7a",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected URL -> downloading to temp file\n",
      "Detected dense format -> using dense reader\n",
      "Temporary file sars.tsv.gz was deleted after loading.\n"
     ]
    }
   ],
   "execution_count": 66
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:09.998928Z",
     "start_time": "2025-07-29T11:43:09.925051Z"
    }
   },
   "cell_type": "code",
   "source": "sars = sars.filter_genes()",
   "id": "5835797041f0e987",
   "outputs": [],
   "execution_count": 67
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Data slots\n",
    "Data is orgnaized in a grandPy object in slots:"
   ],
   "id": "b743f2a2dd97ee90"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:10.034799Z",
     "start_time": "2025-07-29T11:43:10.027461Z"
    }
   },
   "cell_type": "code",
   "source": "sars.slots",
   "id": "fd13687702a765b0",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ntr', 'alpha', 'beta', 'count']"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 68
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "To learn more about metadata, see the [loading data notebook](). After loading GRAND-SLAM analysis results, the default slots are \"count\" (read counts), \"ntr\" (the new-to-total RNA ratio), \"alpha\" and \"beta\" (the parameters for the Beta approximation of the NTR posterior distribution). Each of these slots contains a gene x columns ( columns are either samples or cells, depending on wehther your data is bulk or singel cell data) matrix of numeric values.\n",
    "\n",
    "There is also a default slot, which is used by many functions as default parameter."
   ],
   "id": "23651bd86e1e5bc7"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:10.094753Z",
     "start_time": "2025-07-29T11:43:10.087244Z"
    }
   },
   "cell_type": "code",
   "source": "sars.default_slot",
   "id": "5dc1b4d44d54c326",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'count'"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 69
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "New slots are added by specific grandPy functions such as `normalize()` or `normlaize_tpm()`, which by default, change the default slot. The default slot can also be set manually.",
   "id": "d7f17d0e1f28b293"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:10.186321Z",
     "start_time": "2025-07-29T11:43:10.152873Z"
    }
   },
   "cell_type": "code",
   "source": [
    "sars = sars.normalize()\n",
    "sars.default_slot"
   ],
   "id": "284b5d67d31f51f8",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'norm'"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 70
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:10.230629Z",
     "start_time": "2025-07-29T11:43:10.216966Z"
    }
   },
   "cell_type": "code",
   "source": [
    "sars.with_default_slot(\"count\")\n",
    "sars.default_slot"
   ],
   "id": "f21f5dc9f71f9eaa",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'norm'"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 71
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:10.298270Z",
     "start_time": "2025-07-29T11:43:10.273330Z"
    }
   },
   "cell_type": "code",
   "source": [
    "sars = sars.normalize_tpm(set_to_default = False)\n",
    "sars.default_slot"
   ],
   "id": "3bfaf447ebd36062",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'norm'"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 72
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:10.351321Z",
     "start_time": "2025-07-29T11:43:10.337952Z"
    }
   },
   "cell_type": "code",
   "source": "sars.with_default_slot(\"norm\")\n",
   "id": "77d1bfb80f28b69b",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<GrandPy object: 9162 genes x 12 samples>"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 73
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "There are also other grandPy functions that add additional slots, but do not update the `default_slot` automatically:",
   "id": "6356abb344877152"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:11.123040Z",
     "start_time": "2025-07-29T11:43:10.381016Z"
    }
   },
   "cell_type": "code",
   "source": [
    "sars = sars.compute_ntr_ci()\n",
    "sars.default_slot"
   ],
   "id": "924a72bfd4c1435e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'norm'"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 74
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:11.162279Z",
     "start_time": "2025-07-29T11:43:11.153475Z"
    }
   },
   "cell_type": "code",
   "source": "sars.slots",
   "id": "9f21c6d410983746",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ntr', 'alpha', 'beta', 'count', 'norm', 'tpm', 'lower', 'upper']"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 75
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Analysis\n",
    "In addition to data slots, there is an additional kind of data that is part of a grandPy object: analyses."
   ],
   "id": "d0a27fcbd625372b"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:11.232490Z",
     "start_time": "2025-07-29T11:43:11.225953Z"
    }
   },
   "cell_type": "code",
   "source": "sars.get_analyses()",
   "id": "cdd7bc6296f1a5b3",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 76
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "After loading data there are no analyses, but such data are added e.g. by performing modeling of progressive labeling time courses or analyzing gene expression (see the notebooks [Kinetic modeling]() and [Differential expression]() for more on these)",
   "id": "a12f00482c3a6fa2"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:35.288476Z",
     "start_time": "2025-07-29T11:43:11.298784Z"
    }
   },
   "cell_type": "code",
   "source": [
    "sars = sars.fit_kinetics(name_prefix = \"kinetics\", steady_state = {\"Mock\": True, \"SARS\": False})\n",
    "sars = sars.compute_lfc(contrasts = sars.get_contrasts(contrast = [\"duration.4sU.original\", \"noo4sU\"], group = \"Condition\", no4su = True))\n",
    "sars.analyses"
   ],
   "id": "76e4c6a536ef66e8",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting Mock: 100%|██████████| 9162/9162 [00:11<00:00, 832.62it/s] \n",
      "Fitting SARS: 100%|██████████| 9162/9162 [00:10<00:00, 903.65it/s] \n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'make_name' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001B[31m---------------------------------------------------------------------------\u001B[39m",
      "\u001B[31mNameError\u001B[39m                                 Traceback (most recent call last)",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[77]\u001B[39m\u001B[32m, line 2\u001B[39m\n\u001B[32m      1\u001B[39m sars = sars.fit_kinetics(name_prefix = \u001B[33m\"\u001B[39m\u001B[33mkinetics\u001B[39m\u001B[33m\"\u001B[39m, steady_state = {\u001B[33m\"\u001B[39m\u001B[33mMock\u001B[39m\u001B[33m\"\u001B[39m: \u001B[38;5;28;01mTrue\u001B[39;00m, \u001B[33m\"\u001B[39m\u001B[33mSARS\u001B[39m\u001B[33m\"\u001B[39m: \u001B[38;5;28;01mFalse\u001B[39;00m})\n\u001B[32m----> \u001B[39m\u001B[32m2\u001B[39m sars = sars.compute_lfc(contrasts = \u001B[43msars\u001B[49m\u001B[43m.\u001B[49m\u001B[43mget_contrasts\u001B[49m\u001B[43m(\u001B[49m\u001B[43mcontrast\u001B[49m\u001B[43m \u001B[49m\u001B[43m=\u001B[49m\u001B[43m \u001B[49m\u001B[43m[\u001B[49m\u001B[33;43m\"\u001B[39;49m\u001B[33;43mduration.4sU.original\u001B[39;49m\u001B[33;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[33;43m\"\u001B[39;49m\u001B[33;43mnoo4sU\u001B[39;49m\u001B[33;43m\"\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mgroup\u001B[49m\u001B[43m \u001B[49m\u001B[43m=\u001B[49m\u001B[43m \u001B[49m\u001B[33;43m\"\u001B[39;49m\u001B[33;43mCondition\u001B[39;49m\u001B[33;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mno4su\u001B[49m\u001B[43m \u001B[49m\u001B[43m=\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;28;43;01mTrue\u001B[39;49;00m\u001B[43m)\u001B[49m)\n\u001B[32m      3\u001B[39m sars.analyses\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\PycharmProjects\\grandpy\\grandpy\\grandPy.py:2811\u001B[39m, in \u001B[36mGrandPy.get_contrasts\u001B[39m\u001B[34m(self, contrast, columns, group, name_format, no4su)\u001B[39m\n\u001B[32m   2775\u001B[39m \u001B[38;5;250m\u001B[39m\u001B[33;03m\"\"\"\u001B[39;00m\n\u001B[32m   2776\u001B[39m \u001B[33;03mGenerate contrast matrix for differential comparisons.\u001B[39;00m\n\u001B[32m   2777\u001B[39m \n\u001B[32m   (...)\u001B[39m\u001B[32m   2807\u001B[39m \u001B[33;03m    Values are -1, 0, or 1 depending on contrast design.\u001B[39;00m\n\u001B[32m   2808\u001B[39m \u001B[33;03m\"\"\"\u001B[39;00m\n\u001B[32m   2809\u001B[39m \u001B[38;5;28;01mfrom\u001B[39;00m\u001B[38;5;250m \u001B[39m\u001B[34;01m.\u001B[39;00m\u001B[34;01mdiffexp\u001B[39;00m\u001B[38;5;250m \u001B[39m\u001B[38;5;28;01mimport\u001B[39;00m _get_contrasts\n\u001B[32m-> \u001B[39m\u001B[32m2811\u001B[39m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43m_get_contrasts\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mcontrast\u001B[49m\u001B[43m \u001B[49m\u001B[43m=\u001B[49m\u001B[43m \u001B[49m\u001B[43mcontrast\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mcolumns\u001B[49m\u001B[43m \u001B[49m\u001B[43m=\u001B[49m\u001B[43m \u001B[49m\u001B[43mcolumns\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mgroup\u001B[49m\u001B[43m \u001B[49m\u001B[43m=\u001B[49m\u001B[43m \u001B[49m\u001B[43mgroup\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mname_format\u001B[49m\u001B[43m \u001B[49m\u001B[43m=\u001B[49m\u001B[43m \u001B[49m\u001B[43mname_format\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mno4su\u001B[49m\u001B[43m=\u001B[49m\u001B[43m \u001B[49m\u001B[43mno4su\u001B[49m\u001B[43m)\u001B[49m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\PycharmProjects\\grandpy\\grandpy\\diffexp.py:547\u001B[39m, in \u001B[36m_get_contrasts\u001B[39m\u001B[34m(data, contrast, columns, group, name_format, no4su)\u001B[39m\n\u001B[32m    545\u001B[39m     \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[32m    546\u001B[39m         level_pairs = [(contrast[\u001B[32m1\u001B[39m], contrast[\u001B[32m2\u001B[39m])]\n\u001B[32m--> \u001B[39m\u001B[32m547\u001B[39m     df = \u001B[43mcontrast_df_for_level_pairs\u001B[49m\u001B[43m(\u001B[49m\u001B[43mlevel_pairs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mgroup_name\u001B[49m\u001B[43m=\u001B[49m\u001B[38;5;28;43mstr\u001B[39;49m\u001B[43m(\u001B[49m\u001B[43mgrp_val\u001B[49m\u001B[43m)\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43muse_mask_local\u001B[49m\u001B[43m=\u001B[49m\u001B[43muse_mask_group\u001B[49m\u001B[43m)\u001B[49m\n\u001B[32m    548\u001B[39m     all_dfs.append(df)\n\u001B[32m    549\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m all_dfs:\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\PycharmProjects\\grandpy\\grandpy\\diffexp.py:527\u001B[39m, in \u001B[36m_get_contrasts.<locals>.contrast_df_for_level_pairs\u001B[39m\u001B[34m(level_pairs, group_name, use_mask_local)\u001B[39m\n\u001B[32m    525\u001B[39m \u001B[38;5;28;01mfor\u001B[39;00m a, b \u001B[38;5;129;01min\u001B[39;00m level_pairs:\n\u001B[32m    526\u001B[39m     vec = make_vector(a, b, use_mask_local)\n\u001B[32m--> \u001B[39m\u001B[32m527\u001B[39m     name = \u001B[43mmake_name\u001B[49m(a, b, group_name)\n\u001B[32m    528\u001B[39m     base_data_frame[name] = vec\n\u001B[32m    529\u001B[39m \u001B[38;5;28;01mreturn\u001B[39;00m pd.DataFrame(base_data_frame, index=coldata.index)\n",
      "\u001B[31mNameError\u001B[39m: name 'make_name' is not defined"
     ]
    }
   ],
   "execution_count": 77
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Both analysis methods, `fit_kinetics()` and `compute_lfc()`added multiple analyses: `fit_kinetics()`added an analysis for each `Condition` whereas `compute_lfc()` added an analysis for each of many pairwise comparisons defined by `get_contrasts()` (see [Differential expression]() for details)\n",
    "\n",
    "What is common to data slots and analyses is taht both are tables with as many rows as there are genes. What is different is that the columns of data slots always correspond to the sampels or scells (depending on whether data are bulk or singel cell data), and the columns of analysis tables are arbitrary and depend on the kind of analysis performde.\n",
    "\n",
    "Analysis columns can be retrieved by setting the description parameter to True for `get_analyses`:"
   ],
   "id": "2dfa82e827c34c8d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_analyses(description = True)",
   "id": "7a1de1dd406c8350"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "We see that the `fit_kinetics()` function by default creates tables with two columns (`snthesis` and `half-life`) corresponding to the synthesis rate and RNA half-life for each gene, and the `compute_lfc()` function creates a single column called `lfc` corresponding to the log2 fold cahnges for each gene.",
   "id": "481c0c52ec839d15"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Retrieving data from slots or analyses\n",
    "There are essentially three functions you can use for retrieving slot data:\n",
    "- `get_table()`: The swiss army knife, return a data frame with genes as rows and columns made from potentially several slots and/or analyses; usually for all or at least a lot of genes\n",
    "- `get_data()`: Returns a data frame with the samples or cells as rows and slot data for particular genes in columns; usually for a single or at most very few genes\n",
    "- `get_analysis_table()`: Returns a data frame with genes as rows and columns made from potentially several analyses; usually for all or at least a lot of genes ; there's (almost) no need to call this function (see below for exceptions)\n",
    "\n",
    "# get_table\n",
    "Without any other parameters `get_table()` returns for all genes from the default slot:"
   ],
   "id": "26ea3f4185aae922"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:35.303807300Z",
     "start_time": "2025-07-29T09:52:33.056766Z"
    }
   },
   "cell_type": "code",
   "source": "sars.get_table(mode_slot = \"count\")",
   "id": "b0968d716b2b66df",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "          Mock.no4sU.A   Mock.1h.A   Mock.2h.A   Mock.2h.B  Mock.3h.A  \\\n",
       "Symbol                                                                  \n",
       "MIB2           14.0000    420.0000    372.0000    230.0000     456.50   \n",
       "OSBPL9        133.6667   1202.0000    942.5000   1071.5000    1688.00   \n",
       "BTF3L4        160.3333   1208.1667    738.7500   1251.4167    2071.75   \n",
       "ZFYVE9         51.5000    486.0000    383.3333    364.5000     503.50   \n",
       "PRPF38A        99.5000    938.0000    818.0000    837.0000    1296.00   \n",
       "...                ...         ...         ...         ...        ...   \n",
       "MECP2         187.0000   1925.5000   1624.5000   1318.5000    2492.00   \n",
       "FLNA         5257.5000  67906.3333  64259.3500  45429.1667   82645.60   \n",
       "DNASE1L1       67.0000    758.5000    683.5000    584.5000    1063.25   \n",
       "ORF1ab        459.0000   4421.0000   4567.0000   3378.0000    5459.50   \n",
       "S             267.0000   2530.0000   2657.0000   1901.5000    3062.00   \n",
       "\n",
       "          Mock.4h.A  SARS.no4sU.A    SARS.1h.A     SARS.2h.A     SARS.2h.B  \\\n",
       "Symbol                                                                       \n",
       "MIB2          419.5       14.0000     180.0000  2.650000e+02  2.130000e+02   \n",
       "OSBPL9       1385.5       38.0000     348.0000  3.840000e+02  4.885000e+02   \n",
       "BTF3L4       1371.0       40.0000     369.1667  2.596667e+02  5.014167e+02   \n",
       "ZFYVE9        433.5       19.0000     165.0000  1.620000e+02  2.365000e+02   \n",
       "PRPF38A      1191.0       40.0000     238.0000  2.980000e+02  4.105000e+02   \n",
       "...             ...           ...          ...           ...           ...   \n",
       "MECP2        2283.5       79.0000     618.0000  8.215000e+02  8.080000e+02   \n",
       "FLNA        73480.4     1079.5000   13502.5000  1.668567e+04  1.389320e+04   \n",
       "DNASE1L1      993.5       16.0000     213.0000  2.410000e+02  2.267500e+02   \n",
       "ORF1ab       4486.0   395071.6429  536923.5833  1.351206e+06  2.337993e+06   \n",
       "S            2517.0   271587.3929  286685.2500  7.811136e+05  1.300599e+06   \n",
       "\n",
       "             SARS.3h.A     SARS.4h.A  \n",
       "Symbol                                \n",
       "MIB2      1.000000e+02  1.020000e+02  \n",
       "OSBPL9    1.470000e+02  1.850000e+02  \n",
       "BTF3L4    1.650000e+02  2.120000e+02  \n",
       "ZFYVE9    8.300000e+01  1.080000e+02  \n",
       "PRPF38A   1.360000e+02  1.665000e+02  \n",
       "...                ...           ...  \n",
       "MECP2     3.420000e+02  5.145000e+02  \n",
       "FLNA      7.156667e+03  9.362000e+03  \n",
       "DNASE1L1  9.300000e+01  1.300000e+02  \n",
       "ORF1ab    1.243669e+06  1.775411e+06  \n",
       "S         7.057127e+05  1.046355e+06  \n",
       "\n",
       "[9162 rows x 12 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mock.no4sU.A</th>\n",
       "      <th>Mock.1h.A</th>\n",
       "      <th>Mock.2h.A</th>\n",
       "      <th>Mock.2h.B</th>\n",
       "      <th>Mock.3h.A</th>\n",
       "      <th>Mock.4h.A</th>\n",
       "      <th>SARS.no4sU.A</th>\n",
       "      <th>SARS.1h.A</th>\n",
       "      <th>SARS.2h.A</th>\n",
       "      <th>SARS.2h.B</th>\n",
       "      <th>SARS.3h.A</th>\n",
       "      <th>SARS.4h.A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIB2</th>\n",
       "      <td>14.0000</td>\n",
       "      <td>420.0000</td>\n",
       "      <td>372.0000</td>\n",
       "      <td>230.0000</td>\n",
       "      <td>456.50</td>\n",
       "      <td>419.5</td>\n",
       "      <td>14.0000</td>\n",
       "      <td>180.0000</td>\n",
       "      <td>2.650000e+02</td>\n",
       "      <td>2.130000e+02</td>\n",
       "      <td>1.000000e+02</td>\n",
       "      <td>1.020000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OSBPL9</th>\n",
       "      <td>133.6667</td>\n",
       "      <td>1202.0000</td>\n",
       "      <td>942.5000</td>\n",
       "      <td>1071.5000</td>\n",
       "      <td>1688.00</td>\n",
       "      <td>1385.5</td>\n",
       "      <td>38.0000</td>\n",
       "      <td>348.0000</td>\n",
       "      <td>3.840000e+02</td>\n",
       "      <td>4.885000e+02</td>\n",
       "      <td>1.470000e+02</td>\n",
       "      <td>1.850000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BTF3L4</th>\n",
       "      <td>160.3333</td>\n",
       "      <td>1208.1667</td>\n",
       "      <td>738.7500</td>\n",
       "      <td>1251.4167</td>\n",
       "      <td>2071.75</td>\n",
       "      <td>1371.0</td>\n",
       "      <td>40.0000</td>\n",
       "      <td>369.1667</td>\n",
       "      <td>2.596667e+02</td>\n",
       "      <td>5.014167e+02</td>\n",
       "      <td>1.650000e+02</td>\n",
       "      <td>2.120000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFYVE9</th>\n",
       "      <td>51.5000</td>\n",
       "      <td>486.0000</td>\n",
       "      <td>383.3333</td>\n",
       "      <td>364.5000</td>\n",
       "      <td>503.50</td>\n",
       "      <td>433.5</td>\n",
       "      <td>19.0000</td>\n",
       "      <td>165.0000</td>\n",
       "      <td>1.620000e+02</td>\n",
       "      <td>2.365000e+02</td>\n",
       "      <td>8.300000e+01</td>\n",
       "      <td>1.080000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRPF38A</th>\n",
       "      <td>99.5000</td>\n",
       "      <td>938.0000</td>\n",
       "      <td>818.0000</td>\n",
       "      <td>837.0000</td>\n",
       "      <td>1296.00</td>\n",
       "      <td>1191.0</td>\n",
       "      <td>40.0000</td>\n",
       "      <td>238.0000</td>\n",
       "      <td>2.980000e+02</td>\n",
       "      <td>4.105000e+02</td>\n",
       "      <td>1.360000e+02</td>\n",
       "      <td>1.665000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MECP2</th>\n",
       "      <td>187.0000</td>\n",
       "      <td>1925.5000</td>\n",
       "      <td>1624.5000</td>\n",
       "      <td>1318.5000</td>\n",
       "      <td>2492.00</td>\n",
       "      <td>2283.5</td>\n",
       "      <td>79.0000</td>\n",
       "      <td>618.0000</td>\n",
       "      <td>8.215000e+02</td>\n",
       "      <td>8.080000e+02</td>\n",
       "      <td>3.420000e+02</td>\n",
       "      <td>5.145000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLNA</th>\n",
       "      <td>5257.5000</td>\n",
       "      <td>67906.3333</td>\n",
       "      <td>64259.3500</td>\n",
       "      <td>45429.1667</td>\n",
       "      <td>82645.60</td>\n",
       "      <td>73480.4</td>\n",
       "      <td>1079.5000</td>\n",
       "      <td>13502.5000</td>\n",
       "      <td>1.668567e+04</td>\n",
       "      <td>1.389320e+04</td>\n",
       "      <td>7.156667e+03</td>\n",
       "      <td>9.362000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNASE1L1</th>\n",
       "      <td>67.0000</td>\n",
       "      <td>758.5000</td>\n",
       "      <td>683.5000</td>\n",
       "      <td>584.5000</td>\n",
       "      <td>1063.25</td>\n",
       "      <td>993.5</td>\n",
       "      <td>16.0000</td>\n",
       "      <td>213.0000</td>\n",
       "      <td>2.410000e+02</td>\n",
       "      <td>2.267500e+02</td>\n",
       "      <td>9.300000e+01</td>\n",
       "      <td>1.300000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1ab</th>\n",
       "      <td>459.0000</td>\n",
       "      <td>4421.0000</td>\n",
       "      <td>4567.0000</td>\n",
       "      <td>3378.0000</td>\n",
       "      <td>5459.50</td>\n",
       "      <td>4486.0</td>\n",
       "      <td>395071.6429</td>\n",
       "      <td>536923.5833</td>\n",
       "      <td>1.351206e+06</td>\n",
       "      <td>2.337993e+06</td>\n",
       "      <td>1.243669e+06</td>\n",
       "      <td>1.775411e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>267.0000</td>\n",
       "      <td>2530.0000</td>\n",
       "      <td>2657.0000</td>\n",
       "      <td>1901.5000</td>\n",
       "      <td>3062.00</td>\n",
       "      <td>2517.0</td>\n",
       "      <td>271587.3929</td>\n",
       "      <td>286685.2500</td>\n",
       "      <td>7.811136e+05</td>\n",
       "      <td>1.300599e+06</td>\n",
       "      <td>7.057127e+05</td>\n",
       "      <td>1.046355e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9162 rows × 12 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 40
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "You can change the slot by specifying another `mode_slot`parameter:",
   "id": "8e82d2cecc2c27b7"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:35.303807300Z",
     "start_time": "2025-07-29T09:53:56.483880Z"
    }
   },
   "cell_type": "code",
   "source": "sars.get_table(mode_slot = [\"norm\", \"count\"]).columns",
   "id": "6ec0da88d8fb3426",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Mock.no4sU.A_norm', 'Mock.1h.A_norm', 'Mock.2h.A_norm',\n",
       "       'Mock.2h.B_norm', 'Mock.3h.A_norm', 'Mock.4h.A_norm',\n",
       "       'SARS.no4sU.A_norm', 'SARS.1h.A_norm', 'SARS.2h.A_norm',\n",
       "       'SARS.2h.B_norm', 'SARS.3h.A_norm', 'SARS.4h.A_norm',\n",
       "       'Mock.no4sU.A_count', 'Mock.1h.A_count', 'Mock.2h.A_count',\n",
       "       'Mock.2h.B_count', 'Mock.3h.A_count', 'Mock.4h.A_count',\n",
       "       'SARS.no4sU.A_count', 'SARS.1h.A_count', 'SARS.2h.A_count',\n",
       "       'SARS.2h.B_count', 'SARS.3h.A_count', 'SARS.4h.A_count'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 42
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "By using the `mode_slot` syntax (mode being either of `total`, `new` and `old`), you can also retrieve new RNA counts or new RNA normalized values:",
   "id": "6b96add02c6febed"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:43:35.303807300Z",
     "start_time": "2025-07-29T10:00:19.028690Z"
    }
   },
   "cell_type": "code",
   "source": "sars.get_table(mode_slot= gp.ModeSlot(\"new\", \"norm\"), ntr_nan = True)",
   "id": "7f5ab4c851c4cff4",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "          Mock.no4sU.A    Mock.1h.A    Mock.2h.A    Mock.2h.B    Mock.3h.A  \\\n",
       "Symbol                                                                       \n",
       "MIB2               NaN     2.334512    32.111821    16.474117    40.421769   \n",
       "OSBPL9             NaN    17.339177    55.206225    62.407300    82.942040   \n",
       "BTF3L4             NaN    17.308215    69.380310   120.949589   177.434594   \n",
       "ZFYVE9             NaN     2.267216    27.897675    25.821868    39.302175   \n",
       "PRPF38A            NaN    28.737657   122.185053   133.987687   160.389371   \n",
       "...                ...          ...          ...          ...          ...   \n",
       "MECP2              NaN    50.773766   321.335986   236.242873   398.240228   \n",
       "FLNA               NaN   199.957557  2542.705633  1631.919535  2865.773129   \n",
       "DNASE1L1           NaN     6.399312    53.361958    27.137565    52.135871   \n",
       "ORF1ab             NaN  1125.408603  1223.190926   957.804640   906.522581   \n",
       "S                  NaN   668.059841   725.699918   581.763476   535.948242   \n",
       "\n",
       "            Mock.4h.A  SARS.no4sU.A      SARS.1h.A      SARS.2h.A  \\\n",
       "Symbol                                                              \n",
       "MIB2        52.710641           NaN      10.323628      68.111814   \n",
       "OSBPL9     109.840247           NaN      49.245814     154.332663   \n",
       "BTF3L4     223.605445           NaN      40.315118     101.825258   \n",
       "ZFYVE9      68.879675           NaN       3.283195      56.454819   \n",
       "PRPF38A    252.489814           NaN      82.290788     223.122015   \n",
       "...               ...           ...            ...            ...   \n",
       "MECP2      552.906767           NaN     108.141762     621.657322   \n",
       "FLNA      3817.188192           NaN     284.479165    5490.437208   \n",
       "DNASE1L1    80.851752           NaN       6.232803      55.731343   \n",
       "ORF1ab     844.139520           NaN  224548.040766  995399.866618   \n",
       "S          480.607548           NaN  128250.718166  606114.438989   \n",
       "\n",
       "             SARS.2h.B     SARS.3h.A     SARS.4h.A  \n",
       "Symbol                                              \n",
       "MIB2      3.900246e+01  1.107407e+02  1.165965e+02  \n",
       "OSBPL9    1.816013e+02  1.692269e+02  2.246368e+02  \n",
       "BTF3L4    1.775885e+02  1.949572e+02  2.326411e+02  \n",
       "ZFYVE9    6.681548e+01  1.009159e+02  1.443122e+02  \n",
       "PRPF38A   2.359033e+02  2.318516e+02  2.852863e+02  \n",
       "...                ...           ...           ...  \n",
       "MECP2     4.684457e+02  7.809246e+02  9.330807e+02  \n",
       "FLNA      3.155187e+03  7.761168e+03  9.774823e+03  \n",
       "DNASE1L1  4.752973e+01  4.805717e+01  7.580657e+01  \n",
       "ORF1ab    1.458746e+06  2.341523e+06  2.919400e+06  \n",
       "S         8.439801e+05  1.364170e+06  1.753270e+06  \n",
       "\n",
       "[9162 rows x 12 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mock.no4sU.A</th>\n",
       "      <th>Mock.1h.A</th>\n",
       "      <th>Mock.2h.A</th>\n",
       "      <th>Mock.2h.B</th>\n",
       "      <th>Mock.3h.A</th>\n",
       "      <th>Mock.4h.A</th>\n",
       "      <th>SARS.no4sU.A</th>\n",
       "      <th>SARS.1h.A</th>\n",
       "      <th>SARS.2h.A</th>\n",
       "      <th>SARS.2h.B</th>\n",
       "      <th>SARS.3h.A</th>\n",
       "      <th>SARS.4h.A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIB2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.334512</td>\n",
       "      <td>32.111821</td>\n",
       "      <td>16.474117</td>\n",
       "      <td>40.421769</td>\n",
       "      <td>52.710641</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.323628</td>\n",
       "      <td>68.111814</td>\n",
       "      <td>3.900246e+01</td>\n",
       "      <td>1.107407e+02</td>\n",
       "      <td>1.165965e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OSBPL9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>17.339177</td>\n",
       "      <td>55.206225</td>\n",
       "      <td>62.407300</td>\n",
       "      <td>82.942040</td>\n",
       "      <td>109.840247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>49.245814</td>\n",
       "      <td>154.332663</td>\n",
       "      <td>1.816013e+02</td>\n",
       "      <td>1.692269e+02</td>\n",
       "      <td>2.246368e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BTF3L4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>17.308215</td>\n",
       "      <td>69.380310</td>\n",
       "      <td>120.949589</td>\n",
       "      <td>177.434594</td>\n",
       "      <td>223.605445</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40.315118</td>\n",
       "      <td>101.825258</td>\n",
       "      <td>1.775885e+02</td>\n",
       "      <td>1.949572e+02</td>\n",
       "      <td>2.326411e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFYVE9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.267216</td>\n",
       "      <td>27.897675</td>\n",
       "      <td>25.821868</td>\n",
       "      <td>39.302175</td>\n",
       "      <td>68.879675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.283195</td>\n",
       "      <td>56.454819</td>\n",
       "      <td>6.681548e+01</td>\n",
       "      <td>1.009159e+02</td>\n",
       "      <td>1.443122e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRPF38A</th>\n",
       "      <td>NaN</td>\n",
       "      <td>28.737657</td>\n",
       "      <td>122.185053</td>\n",
       "      <td>133.987687</td>\n",
       "      <td>160.389371</td>\n",
       "      <td>252.489814</td>\n",
       "      <td>NaN</td>\n",
       "      <td>82.290788</td>\n",
       "      <td>223.122015</td>\n",
       "      <td>2.359033e+02</td>\n",
       "      <td>2.318516e+02</td>\n",
       "      <td>2.852863e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MECP2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>50.773766</td>\n",
       "      <td>321.335986</td>\n",
       "      <td>236.242873</td>\n",
       "      <td>398.240228</td>\n",
       "      <td>552.906767</td>\n",
       "      <td>NaN</td>\n",
       "      <td>108.141762</td>\n",
       "      <td>621.657322</td>\n",
       "      <td>4.684457e+02</td>\n",
       "      <td>7.809246e+02</td>\n",
       "      <td>9.330807e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>199.957557</td>\n",
       "      <td>2542.705633</td>\n",
       "      <td>1631.919535</td>\n",
       "      <td>2865.773129</td>\n",
       "      <td>3817.188192</td>\n",
       "      <td>NaN</td>\n",
       "      <td>284.479165</td>\n",
       "      <td>5490.437208</td>\n",
       "      <td>3.155187e+03</td>\n",
       "      <td>7.761168e+03</td>\n",
       "      <td>9.774823e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNASE1L1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>6.399312</td>\n",
       "      <td>53.361958</td>\n",
       "      <td>27.137565</td>\n",
       "      <td>52.135871</td>\n",
       "      <td>80.851752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.232803</td>\n",
       "      <td>55.731343</td>\n",
       "      <td>4.752973e+01</td>\n",
       "      <td>4.805717e+01</td>\n",
       "      <td>7.580657e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1ab</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1125.408603</td>\n",
       "      <td>1223.190926</td>\n",
       "      <td>957.804640</td>\n",
       "      <td>906.522581</td>\n",
       "      <td>844.139520</td>\n",
       "      <td>NaN</td>\n",
       "      <td>224548.040766</td>\n",
       "      <td>995399.866618</td>\n",
       "      <td>1.458746e+06</td>\n",
       "      <td>2.341523e+06</td>\n",
       "      <td>2.919400e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>668.059841</td>\n",
       "      <td>725.699918</td>\n",
       "      <td>581.763476</td>\n",
       "      <td>535.948242</td>\n",
       "      <td>480.607548</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128250.718166</td>\n",
       "      <td>606114.438989</td>\n",
       "      <td>8.439801e+05</td>\n",
       "      <td>1.364170e+06</td>\n",
       "      <td>1.753270e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9162 rows × 12 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 52
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Note that the no4sU columns only have NaN values. You can change this behaviour by specifying the `ntr_nan` parameter:",
   "id": "d9347176765fffcf"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:44:47.090051Z",
     "start_time": "2025-07-29T11:44:47.047389Z"
    }
   },
   "cell_type": "code",
   "source": "sars.get_table(mode_slot= gp.ModeSlot(\"new\", \"norm\"), ntr_nan = False)",
   "id": "e307b16289a487a1",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "          Mock.no4sU.A    Mock.1h.A    Mock.2h.A    Mock.2h.B    Mock.3h.A  \\\n",
       "Symbol                                                                       \n",
       "MIB2               0.0     2.334512    32.111821    16.474117    40.421769   \n",
       "OSBPL9             0.0    17.339177    55.206225    62.407300    82.942040   \n",
       "BTF3L4             0.0    17.308215    69.380310   120.949589   177.434594   \n",
       "ZFYVE9             0.0     2.267216    27.897675    25.821868    39.302175   \n",
       "PRPF38A            0.0    28.737657   122.185053   133.987687   160.389371   \n",
       "...                ...          ...          ...          ...          ...   \n",
       "MECP2              0.0    50.773766   321.335986   236.242873   398.240228   \n",
       "FLNA               0.0   199.957557  2542.705633  1631.919535  2865.773129   \n",
       "DNASE1L1           0.0     6.399312    53.361958    27.137565    52.135871   \n",
       "ORF1ab             0.0  1125.408603  1223.190926   957.804640   906.522581   \n",
       "S                  0.0   668.059841   725.699918   581.763476   535.948242   \n",
       "\n",
       "            Mock.4h.A  SARS.no4sU.A      SARS.1h.A      SARS.2h.A  \\\n",
       "Symbol                                                              \n",
       "MIB2        52.710641           0.0      10.323628      68.111814   \n",
       "OSBPL9     109.840247           0.0      49.245814     154.332663   \n",
       "BTF3L4     223.605445           0.0      40.315118     101.825258   \n",
       "ZFYVE9      68.879675           0.0       3.283195      56.454819   \n",
       "PRPF38A    252.489814           0.0      82.290788     223.122015   \n",
       "...               ...           ...            ...            ...   \n",
       "MECP2      552.906767           0.0     108.141762     621.657322   \n",
       "FLNA      3817.188192           0.0     284.479165    5490.437208   \n",
       "DNASE1L1    80.851752           0.0       6.232803      55.731343   \n",
       "ORF1ab     844.139520           0.0  224548.040766  995399.866618   \n",
       "S          480.607548           0.0  128250.718166  606114.438989   \n",
       "\n",
       "             SARS.2h.B     SARS.3h.A     SARS.4h.A  \n",
       "Symbol                                              \n",
       "MIB2      3.900246e+01  1.107407e+02  1.165965e+02  \n",
       "OSBPL9    1.816013e+02  1.692269e+02  2.246368e+02  \n",
       "BTF3L4    1.775885e+02  1.949572e+02  2.326411e+02  \n",
       "ZFYVE9    6.681548e+01  1.009159e+02  1.443122e+02  \n",
       "PRPF38A   2.359033e+02  2.318516e+02  2.852863e+02  \n",
       "...                ...           ...           ...  \n",
       "MECP2     4.684457e+02  7.809246e+02  9.330807e+02  \n",
       "FLNA      3.155187e+03  7.761168e+03  9.774823e+03  \n",
       "DNASE1L1  4.752973e+01  4.805717e+01  7.580657e+01  \n",
       "ORF1ab    1.458746e+06  2.341523e+06  2.919400e+06  \n",
       "S         8.439801e+05  1.364170e+06  1.753270e+06  \n",
       "\n",
       "[9162 rows x 12 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mock.no4sU.A</th>\n",
       "      <th>Mock.1h.A</th>\n",
       "      <th>Mock.2h.A</th>\n",
       "      <th>Mock.2h.B</th>\n",
       "      <th>Mock.3h.A</th>\n",
       "      <th>Mock.4h.A</th>\n",
       "      <th>SARS.no4sU.A</th>\n",
       "      <th>SARS.1h.A</th>\n",
       "      <th>SARS.2h.A</th>\n",
       "      <th>SARS.2h.B</th>\n",
       "      <th>SARS.3h.A</th>\n",
       "      <th>SARS.4h.A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIB2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.334512</td>\n",
       "      <td>32.111821</td>\n",
       "      <td>16.474117</td>\n",
       "      <td>40.421769</td>\n",
       "      <td>52.710641</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.323628</td>\n",
       "      <td>68.111814</td>\n",
       "      <td>3.900246e+01</td>\n",
       "      <td>1.107407e+02</td>\n",
       "      <td>1.165965e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OSBPL9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>17.339177</td>\n",
       "      <td>55.206225</td>\n",
       "      <td>62.407300</td>\n",
       "      <td>82.942040</td>\n",
       "      <td>109.840247</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.245814</td>\n",
       "      <td>154.332663</td>\n",
       "      <td>1.816013e+02</td>\n",
       "      <td>1.692269e+02</td>\n",
       "      <td>2.246368e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BTF3L4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>17.308215</td>\n",
       "      <td>69.380310</td>\n",
       "      <td>120.949589</td>\n",
       "      <td>177.434594</td>\n",
       "      <td>223.605445</td>\n",
       "      <td>0.0</td>\n",
       "      <td>40.315118</td>\n",
       "      <td>101.825258</td>\n",
       "      <td>1.775885e+02</td>\n",
       "      <td>1.949572e+02</td>\n",
       "      <td>2.326411e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFYVE9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.267216</td>\n",
       "      <td>27.897675</td>\n",
       "      <td>25.821868</td>\n",
       "      <td>39.302175</td>\n",
       "      <td>68.879675</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.283195</td>\n",
       "      <td>56.454819</td>\n",
       "      <td>6.681548e+01</td>\n",
       "      <td>1.009159e+02</td>\n",
       "      <td>1.443122e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRPF38A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>28.737657</td>\n",
       "      <td>122.185053</td>\n",
       "      <td>133.987687</td>\n",
       "      <td>160.389371</td>\n",
       "      <td>252.489814</td>\n",
       "      <td>0.0</td>\n",
       "      <td>82.290788</td>\n",
       "      <td>223.122015</td>\n",
       "      <td>2.359033e+02</td>\n",
       "      <td>2.318516e+02</td>\n",
       "      <td>2.852863e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MECP2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>50.773766</td>\n",
       "      <td>321.335986</td>\n",
       "      <td>236.242873</td>\n",
       "      <td>398.240228</td>\n",
       "      <td>552.906767</td>\n",
       "      <td>0.0</td>\n",
       "      <td>108.141762</td>\n",
       "      <td>621.657322</td>\n",
       "      <td>4.684457e+02</td>\n",
       "      <td>7.809246e+02</td>\n",
       "      <td>9.330807e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLNA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>199.957557</td>\n",
       "      <td>2542.705633</td>\n",
       "      <td>1631.919535</td>\n",
       "      <td>2865.773129</td>\n",
       "      <td>3817.188192</td>\n",
       "      <td>0.0</td>\n",
       "      <td>284.479165</td>\n",
       "      <td>5490.437208</td>\n",
       "      <td>3.155187e+03</td>\n",
       "      <td>7.761168e+03</td>\n",
       "      <td>9.774823e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNASE1L1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.399312</td>\n",
       "      <td>53.361958</td>\n",
       "      <td>27.137565</td>\n",
       "      <td>52.135871</td>\n",
       "      <td>80.851752</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.232803</td>\n",
       "      <td>55.731343</td>\n",
       "      <td>4.752973e+01</td>\n",
       "      <td>4.805717e+01</td>\n",
       "      <td>7.580657e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1ab</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1125.408603</td>\n",
       "      <td>1223.190926</td>\n",
       "      <td>957.804640</td>\n",
       "      <td>906.522581</td>\n",
       "      <td>844.139520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>224548.040766</td>\n",
       "      <td>995399.866618</td>\n",
       "      <td>1.458746e+06</td>\n",
       "      <td>2.341523e+06</td>\n",
       "      <td>2.919400e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.0</td>\n",
       "      <td>668.059841</td>\n",
       "      <td>725.699918</td>\n",
       "      <td>581.763476</td>\n",
       "      <td>535.948242</td>\n",
       "      <td>480.607548</td>\n",
       "      <td>0.0</td>\n",
       "      <td>128250.718166</td>\n",
       "      <td>606114.438989</td>\n",
       "      <td>8.439801e+05</td>\n",
       "      <td>1.364170e+06</td>\n",
       "      <td>1.753270e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9162 rows × 12 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 78
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "`get_table()` can also be used to retrieve analysis results:",
   "id": "99d634a267067a92"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-29T11:44:49.905142Z",
     "start_time": "2025-07-29T11:44:49.786829Z"
    }
   },
   "cell_type": "code",
   "source": "sars.get_table(mode_slot= \"kinetics\")",
   "id": "5259041a41d6ec0b",
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Slot 'kinetics' not found in data slots.",
     "output_type": "error",
     "traceback": [
      "\u001B[31m---------------------------------------------------------------------------\u001B[39m",
      "\u001B[31mValueError\u001B[39m                                Traceback (most recent call last)",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[79]\u001B[39m\u001B[32m, line 1\u001B[39m\n\u001B[32m----> \u001B[39m\u001B[32m1\u001B[39m \u001B[43msars\u001B[49m\u001B[43m.\u001B[49m\u001B[43mget_table\u001B[49m\u001B[43m(\u001B[49m\u001B[43mmode_slot\u001B[49m\u001B[43m=\u001B[49m\u001B[43m \u001B[49m\u001B[33;43m\"\u001B[39;49m\u001B[33;43mkinetics\u001B[39;49m\u001B[33;43m\"\u001B[39;49m\u001B[43m)\u001B[49m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\PycharmProjects\\grandpy\\grandpy\\grandPy.py:1868\u001B[39m, in \u001B[36mGrandPy.get_table\u001B[39m\u001B[34m(self, mode_slot, genes, columns, with_gene_info, name_genes_by, summarize, prefix, ntr_nan, reorder_columns)\u001B[39m\n\u001B[32m   1865\u001B[39m result_df = pd.DataFrame()\n\u001B[32m   1867\u001B[39m \u001B[38;5;28;01mfor\u001B[39;00m slot_name \u001B[38;5;129;01min\u001B[39;00m mode_slot:\n\u001B[32m-> \u001B[39m\u001B[32m1868\u001B[39m     all_data = \u001B[38;5;28;43mself\u001B[39;49m\u001B[43m.\u001B[49m\u001B[43m__resolve_mode_slot\u001B[49m\u001B[43m(\u001B[49m\u001B[43mslot_name\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mntr_nan\u001B[49m\u001B[43m=\u001B[49m\u001B[43mntr_nan\u001B[49m\u001B[43m)\u001B[49m\n\u001B[32m   1870\u001B[39m     \u001B[38;5;28;01mif\u001B[39;00m summarize \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[32m   1871\u001B[39m         matrix = all_data @ summarize.values\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\PycharmProjects\\grandpy\\grandpy\\grandPy.py:472\u001B[39m, in \u001B[36mGrandPy.__resolve_mode_slot\u001B[39m\u001B[34m(self, mode_slot, allow_ntr, ntr_nan)\u001B[39m\n\u001B[32m    450\u001B[39m \u001B[38;5;28;01mdef\u001B[39;00m\u001B[38;5;250m \u001B[39m\u001B[34m__resolve_mode_slot\u001B[39m(\u001B[38;5;28mself\u001B[39m, mode_slot: Union[\u001B[38;5;28mstr\u001B[39m, ModeSlot], *, allow_ntr: \u001B[38;5;28mbool\u001B[39m = \u001B[38;5;28;01mTrue\u001B[39;00m, ntr_nan: \u001B[38;5;28mbool\u001B[39m = \u001B[38;5;28;01mFalse\u001B[39;00m) -> Union[np.ndarray, sp.csr_matrix]:\n\u001B[32m    451\u001B[39m \u001B[38;5;250m    \u001B[39m\u001B[33;03m\"\"\"\u001B[39;00m\n\u001B[32m    452\u001B[39m \u001B[33;03m    Checks whether the given slot is valid and computes the resulting mode slot if a mode was specified.\u001B[39;00m\n\u001B[32m    453\u001B[39m \n\u001B[32m   (...)\u001B[39m\u001B[32m    470\u001B[39m \u001B[33;03m        The resulting slot after the mode has been applied.\u001B[39;00m\n\u001B[32m    471\u001B[39m \u001B[33;03m    \"\"\"\u001B[39;00m\n\u001B[32m--> \u001B[39m\u001B[32m472\u001B[39m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43mself\u001B[39;49m\u001B[43m.\u001B[49m\u001B[43m__slot_tool\u001B[49m\u001B[43m.\u001B[49m\u001B[43mresolve_mode_slot\u001B[49m\u001B[43m(\u001B[49m\u001B[43mmode_slot\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mallow_ntr\u001B[49m\u001B[43m=\u001B[49m\u001B[43mallow_ntr\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mntr_nan\u001B[49m\u001B[43m=\u001B[49m\u001B[43mntr_nan\u001B[49m\u001B[43m)\u001B[49m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\PycharmProjects\\grandpy\\grandpy\\slot_tool.py:180\u001B[39m, in \u001B[36mSlotTool.resolve_mode_slot\u001B[39m\u001B[34m(self, mode_slot, allow_ntr, ntr_nan)\u001B[39m\n\u001B[32m    177\u001B[39m mode_slot = _parse_as_mode_slot(mode_slot)\n\u001B[32m    179\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28mself\u001B[39m.check_slot(mode_slot.slot, allow_ntr=allow_ntr):\n\u001B[32m--> \u001B[39m\u001B[32m180\u001B[39m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mValueError\u001B[39;00m(\u001B[33mf\u001B[39m\u001B[33m\"\u001B[39m\u001B[33mSlot \u001B[39m\u001B[33m'\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mmode_slot.slot\u001B[38;5;132;01m}\u001B[39;00m\u001B[33m'\u001B[39m\u001B[33m not found in data slots.\u001B[39m\u001B[33m\"\u001B[39m)\n\u001B[32m    182\u001B[39m slot_total = \u001B[38;5;28mself\u001B[39m._adata.layers[mode_slot.slot].copy()\n\u001B[32m    183\u001B[39m ntr = \u001B[38;5;28mself\u001B[39m._adata.layers.get(\u001B[33m\"\u001B[39m\u001B[33mntr\u001B[39m\u001B[33m\"\u001B[39m)\n",
      "\u001B[31mValueError\u001B[39m: Slot 'kinetics' not found in data slots."
     ]
    }
   ],
   "execution_count": 79
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Note that you can also specify the full name (it actually is a regular expression that is matched against each analysis name).\n",
    "\n",
    "It is also easily possible to only retrieve data for specific columns (i.e., samples or cells) by using the `columns` parameter. Note that you can use names from the `coldata` table to construct a logical vector over the columns; using a character vector (to specify names) or a numeric vector (to specify positions) also works:"
   ],
   "id": "911843e58ef9842d"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
