{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Kinetic modeling\n",
    "\n",
    "One of the main applications of metabolic RNA labeling is to model the kinetics of RNA expression levels [[Narain, Ashwin et al. (2021)]](https://www.cell.com/molecular-cell/fulltext/S1097-2765(21)00496-2). The most natural and widely used model is to describe the change of RNA levels $a(t)$ and time $t$ is by the differential equation:\n",
    "\n",
    "$$\\frac{da}{dt} = \\sigma - \\delta \\cdot a(t)$$\n",
    "\n",
    "Here $\\sigma$ is the net synthesis rate of RNA and $\\delta$ is the degradation rate. This differential equation can be solved analytically (see e.g. [Jürges et al. (2018)](https://academic.oup.com/bioinformatics/article/34/13/i218/5045735)). Based on this, there are several ways implemented in GrandPy to estimate both $\\sigma$ and $\\delta$. Here, we will see how to fit this model using non-linear least squares (NLLS) regression on the estimated new and old RNA counts [[Narain, Ashwin et al. (2021)]](https://www.cell.com/molecular-cell/fulltext/S1097-2765(21)00496-2).\n",
    "\n",
    "To perform an analysis of the RNA dynamics for data from [Finkel et al. (2021)](https://www.nature.com/articles/s41586-021-03610-3), we first load the GrandPy package and then read the GRAND-SLAM output table directly from zenodo."
   ],
   "id": "8d6e2823629e5e0d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "import warnings\n",
    "import grandpy as gp\n",
    "from grandpy import plot_pca, plot_gene_progressive_timecourse\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "\n",
    "sars = gp.read_grand(\"https://zenodo.org/record/5834034/files/sars.tsv.gz\", design=(\"Condition\", \"dur.4sU\", \"Replicate\"))"
   ],
   "id": "b1ab5d47d6c1ba9b"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Refer to the [loading data and working with GrandPy objects](../notebook_03_loading_data_and_working_with_grandpy_objects.ipynb) notebook to learn more about how to load data. Note that sample metadata has been automatically extracted from the sample names via the design parameter given to `read_grand()`.",
   "id": "8acdf71ef3964f2a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.coldata",
   "id": "c53f2cd7b0794623"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "By default GRAND-SLAM will report data on all genes (with at least one mapped read) and `read_grand()` will read all these genes from the output:",
   "id": "8924c44efd791a45"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars)",
   "id": "ac5c91e1385dcb30"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Thus, we filter to only include genes that have at least 100 reads in the at least 6 samples:",
   "id": "21f35b9c20be2d60"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.filter_genes(min_expression=100, min_columns=6)\n",
    "print(sars)"
   ],
   "id": "14feeac371e8d83f"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "The actual data is available in `slots`. `read_grand()` adds the read counts, new to total RNA ratios (NTRs) and information on the NTR posterior distribution (alpha, beta).\n",
    "\n",
    "As a quick quality check, we can inspect a principal component analysis of all samples involved:"
   ],
   "id": "ea53f9bb39a88a38"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_pca(sars)",
   "id": "6e3c0d44567e314e"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "The samples are colored according to *Condition* (see above in the Coldata table). *Condition* has a special meaning in GrandPy, not only for `plot_pca()`, but also for other analyses (see below). `condition` can be set conveniently using `with_condition()`. For more information, see the [loading data and working with GrandPy objects](../notebook_03_loading_data_and_working_with_grandpy_objects.ipynb) notebook. For `plot_pca()`, the visual attributes can be adapted via the `aest` parameter:",
   "id": "b6c8c91b56655836"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_pca(sars, aest={\"color\": \"duration.4sU.original\", \"shape\": \"Condition\"})",
   "id": "a25062c4b0066353"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "There are no obvious problems with the samples, even though the virus infected no4sU sample is the bottom-most whereas the other virus infected samples are ordered along infection time from bottom to top.\n",
    "\n",
    "For the NLLS approach it is important to normalize the data:"
   ],
   "id": "c95d74572d8d1d47"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.normalize()\n",
    "print(sars)"
   ],
   "id": "f7a7c9837c6bc239"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Make sure that the normalization you use here is appropriate. Calling `normalize()` will add a slot which is set to be the default slot.\n",
    "\n",
    "Before we estimate kinetic parameters globally, we inspect an example:"
   ],
   "id": "5600b9b89b776466"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_gene_progressive_timecourse(sars, \"SRSF6\")",
   "id": "bab7472a65590e37"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "The curves represent the fitted model for this gene. The kinetic modeling by default makes the assumption of steady state gene expression, i.e. that as much RNA is transcribed per time unit as it is degraded. In mathematical terms, it is\n",
    "\n",
    "$$\\frac{da}{dt} = \\sigma - \\delta \\cdot a(t) \\quad \\Leftrightarrow \\quad a(t) = \\frac{\\sigma}{\\delta}$$\n",
    "\n",
    "For the virus infected samples (“SARS”), this is not the case. So we specify an additional parameter:"
   ],
   "id": "1bd695ffe5f42a50"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_gene_progressive_timecourse(sars, \"SRSF6\", steady_state={\"Mock\": True, \"SARS\": False})",
   "id": "1812df0f7478ffc3"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Note that the fit actually changed. Now we are ready to fit the model for each gene. For that we take a closer look at `fit_kinetics()`.\n",
    "\n",
    "By default, `fit_kinetics()` will run in parallel if the dataset is large enough. This can be controlled via the `max_processes` and `exact_processes` parameter.\n",
    "\n",
    "`max_processes` defines the maximum number of worker processes (typically corresponding to CPU cores) that the method can use. `exact_processes` defines, whether to use exactly `max_processes`. If set to `True`, the specified number of processes will be launched, regardless of the workload. If it is set to `False`, the number of processes is adjusted dynamically based on the system's number of cores and the size of the task (This is a rough estimate and might not be optimal, depending on the specific system).\n",
    "\n",
    "Here we only specify the `steady_state` parameter."
   ],
   "id": "30026f58e9ae8f30"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars = sars.fit_kinetics(steady_state={\"Mock\": True, \"SARS\": False})",
   "id": "e1487bc00a2f40ed"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "`fit_kinetics()` added its result to the GrandPy object.",
   "id": "9ff438d53ec5ef49"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars)",
   "id": "9a870862be66e848"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Note that there are now two analysis tables. We can get more information using `get_analyses()`:",
   "id": "266c9be4b4fc414f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_analyses(description=True)",
   "id": "98344e38ecbf4f65"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "This tells us that the two analysis tables, each have a Synthesis and a Half-life column. We can retrieve the analysis results using the `get_analysis_table()` method:",
   "id": "56d26188ad0b8bda"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_analysis_table().head()",
   "id": "5cc040002781bd0a"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "See the [working with data matrices and analysis results](../notebook_04_working_with_data_matrices_and_analysis_results.ipynb) notebook for more information on how to retrieve data from the GrandPy object.\n",
    "\n",
    "We can use the `plot_scatter()` function to compare half-lifes from 'Mock' and 'SARS'."
   ],
   "id": "f8c49dcb798f636d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "gp.plot_scatter(sars, x=\"Mock_Half-life\", y=\"SARS_Half-life\",log=True, diagonal=True)",
   "id": "5389022c7ec4bd9"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "For more on `plot_scatter()`, see the [plotting notebook](...).`fit_kinetics()` can actually compute much more information. To see this, let's look at the `return_fields` parameter.",
   "id": "d106dbd00d07a375"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "small_sars = sars[0:100] # Only use the first 100 genes for faster computation\n",
    "\n",
    "small_sars = small_sars.fit_kinetics(steady_state={\"Mock\": True, \"SARS\": False}, return_fields=[\"Synthesis\", \"Degradation\", \"Half-life\", \"log_likelihood\", \"f0\", \"total\", \"conf_lower\", \"conf_upper\", \"rmse\"], show_progress=False)\n",
    "\n",
    "print(small_sars.get_analyses(description=True))"
   ],
   "id": "e2a01e4962dde841"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "We could also extract the residuals from the model fit. For this we go back to the entire dataset. Note that our previous fit result will be overwritten. This could be prevented by setting the `name_prefix` parameter.",
   "id": "365afad3218efa07"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.fit_kinetics(steady_state={\"Mock\": True, \"SARS\": False}, return_fields=\"residuals\")\n",
    "print(sars.get_analyses(description=True))"
   ],
   "id": "3245252a02c5f498"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Let's plot the distribution of relative residuals for new RNA.",
   "id": "708cf14008be3f4b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "df = sars.get_analysis_table(columns=[\"relative_new\"], with_gene_info=False)\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "\n",
    "df.columns = df.columns.str.replace('_residuals_relative_new', '')\n",
    "\n",
    "df_melted = pd.melt(df, var_name=\"Sample\", value_name=\"Relative residual\")\n",
    "\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.set_theme(style=\"whitegrid\")\n",
    "\n",
    "sns.boxplot(data=df_melted, x='Sample', y='Relative residual', palette='rainbow')\n",
    "\n",
    "plt.axhline(0, color='black', linewidth=1)\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "plt.ylim(-1.1, 1.1)\n",
    "\n",
    "plt.xlabel('Sample', fontsize=14)\n",
    "plt.ylabel('Relative residual', fontsize=14)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ],
   "id": "5fc7dd861255992d"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "We see that the residuals from the 1h timepoints are systematically below zero. An example of this is 'SMAD3':",
   "id": "5e29f4e992f3123f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_gene_progressive_timecourse(sars, \"SMAD3\", steady_state={\"Mock\": True, \"SARS\": False})",
   "id": "e4f4ba1b6fd35241"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "For both 'Mock' and 'SARS', we see that new RNA from 1h is well below the model fit, i.e. has negative residual. This indicates that the effective labeling time was much lower than the nominal labeling time of 1h for these samples, i.e. that actually, these samples should move to the left. This can be rectified by temporal recalibration:",
   "id": "82bf60fb37165403"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# This takes a long time to compute properly, so we just take the 10 top expressed genes here for a rough estimate.\n",
    "sars = sars.calibrate_effective_labeling_time_kinetic_fit(steady_state={\"Mock\": True, \"SARS\": False}, n_top_genes=10)"
   ],
   "id": "6e1e7625c27d374d"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "This adds the column *calibrated_time* to the `coldata`.",
   "id": "9bc9f556815992dc"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.coldata",
   "id": "318fac5e4c36cf4f"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "This recalibrated time can now be used for fitting:",
   "id": "d0ce622d3cfe0137"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_gene_progressive_timecourse(sars, \"SMAD3\", steady_state={\"Mock\": True, \"SARS\": False}, time=\"calibrated_time\", exact_tics=False)",
   "id": "faf7cede0414d6f4"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars = sars.fit_kinetics(name_prefix=\"corrected\", time=\"calibrated_time\", steady_state={\"Mock\": True, \"SARS\": False}, return_fields=\"residuals\")",
   "id": "ee79206f422eec3e"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "This indeed corrected the residuals:",
   "id": "f8a9cb11fecb34b1"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "df = sars.get_analysis_table(columns=[\"corrected.*relative_new\"], with_gene_info=False)\n",
    "\n",
    "df.columns = df.columns.str.replace('_residuals_relative_new', '')\n",
    "df.columns = df.columns.str.replace('corrected_', '')\n",
    "\n",
    "df_melted = pd.melt(df, var_name=\"Sample\", value_name=\"Relative residual\")\n",
    "\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.set_theme(style=\"whitegrid\")\n",
    "\n",
    "sns.boxplot(data=df_melted, x='Sample', y='Relative residual', palette='rainbow')\n",
    "\n",
    "plt.axhline(0, color='black', linewidth=1)\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "plt.ylim(-1.1, 1.1)\n",
    "\n",
    "plt.xlabel('Sample', fontsize=14)\n",
    "plt.ylabel('Relative residual', fontsize=14)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ],
   "id": "2d8ec929a1f63474"
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
