{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Working with data matrices and analysis results\n",
    "\n",
    "This notebook will show the most suitable commands to retrieve data from GrandPy object in different scenarios.\n",
    "\n",
    "Throughout this notebook, we will be using GRAND-SLAM processed SLAM-seq data set from [Finkel et al. (2021)](https://www.nature.com/articles/s41586-021-03610-3). The data set contains time series (progressive labeling) samples from a human epithelial cell line (Calu3 cells); half of the samples were infedctes with SARS-CoV-2 for different periods of time. For more on these initial commands see the [loading data notebook](../notebook_03_loading_data_and_working_with_grandpy_objects)."
   ],
   "id": "db855490635fbf3b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "import warnings\n",
    "import pandas as pd\n",
    "from scripts.regsetup import description\n",
    "\n",
    "import grandpy as gp\n",
    "from grandpy import ModeSlot\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "\n",
    "sars = gp.read_grand(\"https://zenodo.org/record/5834034/files/sars.tsv.gz\", design=(\"Condition\", \"dur.4sU\", \"Replicate\"), classify_genes_func=lambda df: gp.classify_genes(df, cg_name=\"viral\"))"
   ],
   "id": "a9f1ac577f22d99"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars = sars.filter_genes()",
   "id": "f9f818f0130d3d9f"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Data slots\n",
    "Data is organized in `slots`:"
   ],
   "id": "26f5695b4e24f656"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.slots",
   "id": "a2e8604693f28dee"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "To learn more about metadata, see the [loading data notebook](../notebook_03_loading_data_and_working_with_grandpy_objects). After loading GRAND-SLAM analysis results, the standard slots are \"count\" (read counts), \"ntr\" (the new-to-total RNA ratio), \"alpha\" and \"beta\" (the parameters for the Beta approximation of the NTR posterior distribution). Each of these slots contains a genes x columns matrix of numeric values. (columns are either samples or cells, depending on whether your data is bulk or single cell data)\n",
    "\n",
    "There is also a `default_slot`, which is used by many functions as a default parameter."
   ],
   "id": "d02d42580105e18e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.default_slot",
   "id": "9acd7eac92548cac"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "New slots are added by specific GrandPy methods such as `normalize()` or `normalize_tpm()`, which change the `default_slot`. It can also be set manually.",
   "id": "d0359aa6ea158f45"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.normalize()\n",
    "sars.default_slot"
   ],
   "id": "528874974e36bbcc"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.with_default_slot(\"count\")\n",
    "sars.default_slot"
   ],
   "id": "3c753028904881c5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.normalize_tpm(set_to_default = False)\n",
    "sars.default_slot"
   ],
   "id": "f8174cb120cb5e3e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.with_default_slot(\"norm\")\n",
    "sars.default_slot"
   ],
   "id": "71d5fe119da7188e"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "There are also other GrandPy functions that add additional slots, but do not update the `default_slot` automatically:",
   "id": "82c44d2d21c336e6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sars = sars.compute_ntr_ci()\n",
    "sars.default_slot"
   ],
   "id": "cce0076eb76488ed"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.slots",
   "id": "c8c4241c405fc041"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Analysis\n",
    "In addition to data slots, there is an additional kind of data that is part of a GrandPy object: analyses."
   ],
   "id": "dbf003a04247dc1b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.analyses",
   "id": "1727ffc492894d8f"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "After loading data there are no analyses, but such data is added e.g. by performing modeling of progressive labeling time courses or analyzing gene expression (see the notebooks [kinetic modeling](../notebook_02_kinetic_modeling) and [differential expression](../notebook_01_differential_expression) for more on these)",
   "id": "d6f472fbec31401a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# sars = sars.fit_kinetics(steady_state = {\"Mock\": True, \"SARS\": False})\n",
    "# sars = sars.compute_lfc(contrasts = sars.get_contrasts(contrast = [\"duration.4sU.original\", \"no4sU\"], group = \"Condition\", no4su = True))\n",
    "sars.analyses"
   ],
   "id": "15fdd75a164af153"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Both analysis methods, `fit_kinetics()` and `compute_lfc()` added multiple analyses: `fit_kinetics()` added an analysis for each `condition` whereas `compute_lfc()` added an analysis for each pairwise comparison defined by `get_contrasts()` (see [differential expression](../notebook_01_differential_expression) for details).\n",
    "\n",
    "What is common to data slots and analyses is that both are tables with as many rows as there are genes. The main difference are their columns. In `slots` the columns correspond to the samples or cells (depending on whether data are bulk or single cell data). In `analyses` the columns are arbitrary, only depending on the kind of analysis performed.\n",
    "\n",
    "Analysis columns can be retrieved by setting the `description` to 'True' for `get_analyses()`:"
   ],
   "id": "fd43acf454f57c5e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_analyses(description = True)",
   "id": "526679c523af4794"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "We see that `fit_kinetics()` by default creates tables with two columns (*Synthesis* and *Half-life*) corresponding to the synthesis rate and RNA half-life for each gene. <span style=\"color:red;\">`compute_lfc()` creates a single column called *lfc* corresponding to the log2 fold changes for each gene.",
   "id": "c860f6e5ae557823"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Retrieving data from slots or analyses\n",
    "There are three main methods you can use for retrieving the data from `slots`:\n",
    "- `get_table()`: Returns a DataFrame with genes as rows and columns made from potentially several slots\n",
    "- `get_data()`: Returns a DataFrame with the samples/cells as rows and slot data for particular genes in columns\n",
    "- `get_analysis_table()`: Returns a DataFrame with genes as rows and columns made from potentially several analyses\n",
    "\n",
    "# get_table\n",
    "Without any other parameters `get_table()` returns the data from `default_slot` for all genes:"
   ],
   "id": "5022e103b97f4f53"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table().head()",
   "id": "9afef8bcba8e6ff0"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "You can change the slot by specifying the `mode_slot`.",
   "id": "bc7fd70278c93d9e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(mode_slot = \"count\").head()",
   "id": "6a5eb28f3f268978"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Multiple slots can be used at once:",
   "id": "9336eb4c76dd9cc5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(mode_slot = [\"norm\", \"count\"]).columns",
   "id": "ab985b1596ec14d8"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "By using the `mode_slot` syntax (mode being either of *total*, *new* and *old*), you can also retrieve new RNA counts or new RNA normalized values:",
   "id": "59bad2fca4eace3b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(mode_slot=\"new_norm\").head()",
   "id": "b77ac368f783d22"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "A mode slot can also be specified using a `ModeSlot` object, as you will see in the following example.\n",
    "\n",
    "Note that the no4sU columns contain only NaN values. You can change this behavior by specifying `ntr_nan`:"
   ],
   "id": "231e30225ed41d05"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(mode_slot= gp.ModeSlot(\"new\", \"norm\"), ntr_nan = False).head()",
   "id": "50fe4f43a6945960"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "It is also possible to only retrieve data for specific `columns` (i.e., samples or cells) by using the `columns` parameter. Here you can use column names, their index, or a boolean mask:",
   "id": "6b55fbe3dcc375bd"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(columns = [\"Mock.no4sU.A\", \"SARS.no4sU.A\"]).columns",
   "id": "58ceca02d5990f8d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(columns = sars.columns[3:6]).columns",
   "id": "32fb77dfb7e9d7f7"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(columns = (sars.coldata[\"duration.4sU\"] >= 2) & (sars.coldata[\"Condition\"] == \"Mock\")).columns",
   "id": "498c9b0ec08d4e6"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "It is furthermore possible to only fetch data for specific genes (e.g. viral genes) using the `genes` parameter. It is either gene names/symbols, their index, or a boolean mask:",
   "id": "c235ffa765b4ccd6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(genes = \"MYC\").index",
   "id": "88b1d8ac786bfa56"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(genes = range(0,3)).index",
   "id": "bbfc17a7c5fee4ab"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(genes = (sars.gene_info[\"Type\"] == \"viral\")).index",
   "id": "b964c8a00b4418fb"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Sometimes, it makes sense to add the `gene_info` to the DataFrame (for more on gene metadata, see the [loading data notebook](../notebook_03_loading_data_and_working_with_grandPy_objects)):",
   "id": "9c9a41438914b01a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "df = sars.get_table(mode_slot = \"norm\", with_gene_info = True)\n",
    "df.head()"
   ],
   "id": "ca3828b08ca0c3a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "sns.scatterplot(\n",
    "    data=df,\n",
    "    x='SARS.4h.A',\n",
    "    y='SARS.no4sU.A',\n",
    "    hue='Type'\n",
    ")\n",
    "\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "\n",
    "lims = [\n",
    "    np.min([df['SARS.4h.A'].min(), df['SARS.no4sU.A'].min()]),\n",
    "    np.max([df['SARS.4h.A'].max(), df['SARS.no4sU.A'].max()]),\n",
    "]\n",
    "plt.plot(lims, lims, 'k-')\n",
    "plt.xlim(lims)\n",
    "plt.ylim(lims)\n",
    "plt.xlabel('SARS.4h.A')\n",
    "plt.ylabel('SARS.no4sU.A')\n",
    "plt.title('Log-Log Scatter Plot')\n",
    "plt.legend(title='Type')\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "highlight = sars.get_classified_genes(\"viral\")\n",
    "# gp.plot_scatter(sars, x = \"SARS.4h.A\",y = \"SARS.no4sU.A\",log = True, remove_outlier=False, diagonal=True, color=\"blue\", highlight=highlight)\n",
    "# gp.plot_scatter(sars, x=\"SARS.4h.A\", y=\"SARS.no4sU.A\", remove_outlier=False, log_x=True, log_y=True, diagonal=True)"
   ],
   "id": "a890b8469226cef1"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Finally, values can be summarized across samples or cells from the same `condition`:",
   "id": "3c19d5021c3629c0"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(summarize = sars.get_summarize_matrix()).head()",
   "id": "ff71785d0cca7973"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "This is achieved by a summerize matrix:",
   "id": "b786153dd78bb87a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_summarize_matrix()",
   "id": "a4f226ff3fdd9657"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "For summarization, the summarize matrix is matrix-multiplied with the raw matrix. `get_summarized_matrix()` will generate a matrix with a column for each `condition`:",
   "id": "49f865651f0eed3a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print(sars.condition)",
   "id": "b3b8ae0ff9d14bd8"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "By default, no4sU columns are removed (i.e. zero in the matrix), but the `no4su` parameter can change this:",
   "id": "2a1b001d9215088c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_summarize_matrix(no4su = True)",
   "id": "d33a127cd64571e8"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "It is also possible to only focus on specific columns:",
   "id": "e07c4e0cecc2af83"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_summarize_matrix(columns =sars.coldata[\"duration.4sU\"] < 4)",
   "id": "6e2715dfa2faffa4"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "The default behavior is to compute the average, which can be changed to computing sums:",
   "id": "59691419e4c87f04"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_summarize_matrix(average = False)",
   "id": "28caf1fbb8559f5f"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "As a final example, let's get the averaged normalized values for only the 2h timepoint:",
   "id": "2be498b2b091f19f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_table(summarize=sars.get_summarize_matrix(columns=sars.coldata[\"duration.4sU\"] == 2))",
   "id": "b23f8ccd62632008"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# get_data\n",
    "`get_data()` is very similar to `get_table()`. The most important difference being its shape. The Output is transposed relative to `get_table()`."
   ],
   "id": "cb5735aa94c0ec58"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_data(genes = \"MYC\")",
   "id": "ce6da032f83c24d7"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Note that here the `coldata` can be added, which is also the default behavior (for more on column metadata, see the [loading data notebook](../notebook_03_loading_data_and_working_with_grandPy_objects)). This can be changed by using the `with_coldata` parameter:",
   "id": "6c4a17f64323959e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_data(genes = \"MYC\", with_coldata = False)",
   "id": "a7f44f2528e6c583"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "The `genes`, `columns` and `mode_slot` parameters work just like they do in `get_table()`.",
   "id": "ee04673b825648b7"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_data(mode_slot = [\"count\", \"norm\"], genes = [\"MYC\", \"SRSF6\"], columns = sars.coldata[\"Condition\"] == \"Mock\", with_coldata = False)",
   "id": "434e8797984557cc"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Finally, it is also possible to append multiple genes (and/or slots) not as columns, but as additional rows:",
   "id": "da49dbc8563dbd27"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sars.get_data(genes = [\"MYC\", \"SRSF6\"], columns = sars.coldata[\"duration.4sU\"] < 2, by_rows = True)",
   "id": "a583245c9d3468b"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "This can be quite helpful, as for the following example: We retrieve total, old new RNA for SRSF6 (only replicate A), and do this by rows. This way, the data can directly be used for mathplotlib/seaborn to plot the progressive labeling time course (note the much shorter half-life, which is the time where the new and old lines cross, for SARS as compared to Mock):",
   "id": "20e4b60e6637c20e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "df = sars.get_data(mode_slot = [\"old_norm\", \"new_norm\", \"total_norm\"], genes = \"SRSF6\",\n",
    "                   columns = sars.coldata[\"Replicate\"] == \"A\", by_rows = True, ntr_nan=True)\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 6), sharey=True)\n",
    "\n",
    "sns.lineplot(ax=axes[0], data=df[df['Condition'] == 'Mock'], x='duration.4sU', y='Value',\n",
    "             hue='Slot', style='Slot', markers=True, dashes=False)\n",
    "\n",
    "axes[0].set_title('Mock')\n",
    "axes[0].set_xlabel('duration.4sU')\n",
    "axes[0].set_ylabel('Value')\n",
    "\n",
    "sns.lineplot(ax=axes[1], data=df[df['Condition'] == 'SARS'], x='duration.4sU', y='Value',\n",
    "             hue='Slot', style='Slot', markers=True, dashes=False)\n",
    "\n",
    "axes[1].set_title('SARS')\n",
    "axes[1].set_xlabel('duration.4sU')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ],
   "id": "91356f64eb6461a6"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# get_analysis_table\n",
    "`get_analysis_table` "
   ],
   "id": "74e622d192df369"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
